diff --git a/lib/tsc.js b/lib/tsc.js
index ac6a4ef1c..216a4d39d 100644
semver exclusivity >=5.3.0
--- a/lib/tsc.js
+++ b/lib/tsc.js
@@ -5046,6 +5046,9 @@ var sys = (() => {
       return Buffer.from && Buffer.from !== Int8Array.from ? Buffer.from(input, encoding) : new Buffer(input, encoding);
     }
     function isFileSystemCaseSensitive() {
+      if (typeof process.versions.pnp !== `undefined`) {
+        return true;
+      }
       if (platform === "win32" || platform === "win64") {
         return false;
       }
@@ -5427,12 +5430,19 @@ function getAnyExtensionFromPath(path, extensions, ignoreCase) {
   }
   return "";
 }
+var pathComponentsCache = /* @__PURE__ */ new Map();
 function pathComponents(path, rootLength) {
+  const cached = pathComponentsCache.get(path);
+  if (cached) {
+    return cached.slice();
+  }
   const root = path.substring(0, rootLength);
   const rest = path.substring(rootLength).split(directorySeparator);
   if (rest.length && !lastOrUndefined(rest))
     rest.pop();
-  return [root, ...rest];
+  const res = [root, ...rest];
+  pathComponentsCache.set(path, res);
+  return res;
 }
 function getPathComponents(path, currentDirectory = "") {
   path = combinePaths(currentDirectory, path);
@@ -36782,6 +36792,46 @@ function getDefaultValueForOption(option) {
   }
 }

+// src/compiler/pnp.ts
+function getPnpApi(path) {
+  if (typeof process.versions.pnp === "undefined") {
+    return;
+  }
+  const { findPnpApi } = require("module");
+  if (findPnpApi) {
+    return findPnpApi(`${path}/`);
+  }
+}
+function getPnpApiPath(path) {
+  var _a;
+  return (_a = getPnpApi(path)) == null ? void 0 : _a.resolveRequest(
+    "pnpapi",
+    /*issuer*/
+    null
+  );
+}
+function getPnpTypeRoots(currentDirectory) {
+  const pnpApi = getPnpApi(currentDirectory);
+  if (!pnpApi) {
+    return [];
+  }
+  currentDirectory = resolvePath(currentDirectory);
+  const currentPackage = pnpApi.findPackageLocator(`${currentDirectory}/`);
+  if (!currentPackage) {
+    return [];
+  }
+  const { packageDependencies } = pnpApi.getPackageInformation(currentPackage);
+  const typeRoots = [];
+  for (const [name, referencish] of Array.from(packageDependencies.entries())) {
+    if (name.startsWith(`@types/`) && referencish !== null) {
+      const dependencyLocator = pnpApi.getLocator(name, referencish);
+      const { packageLocation } = pnpApi.getPackageInformation(dependencyLocator);
+      typeRoots.push(getDirectoryPath(packageLocation));
+    }
+  }
+  return typeRoots;
+}
+
 // src/compiler/moduleNameResolver.ts
 function trace(host, message, ...args) {
   host.trace(formatMessage(message, ...args));
@@ -36992,7 +37042,7 @@ function getEffectiveTypeRoots(options, host) {
     return getDefaultTypeRoots(currentDirectory);
   }
 }
-function getDefaultTypeRoots(currentDirectory) {
+function getNodeModulesTypeRoots(currentDirectory) {
   let typeRoots;
   forEachAncestorDirectory(normalizePath(currentDirectory), (directory) => {
     const atTypes = combinePaths(directory, nodeModulesAtTypes);
@@ -37005,6 +37055,15 @@ function arePathsEqual(path1, path2, host) {
   const useCaseSensitiveFileNames2 = typeof host.useCaseSensitiveFileNames === "function" ? host.useCaseSensitiveFileNames() : host.useCaseSensitiveFileNames;
   return comparePaths(path1, path2, !useCaseSensitiveFileNames2) === 0 /* EqualTo */;
 }
+function getDefaultTypeRoots(currentDirectory) {
+  const nmTypes = getNodeModulesTypeRoots(currentDirectory);
+  const pnpTypes = getPnpTypeRoots(currentDirectory);
+  if (nmTypes == null ? void 0 : nmTypes.length) {
+    return [...nmTypes, ...pnpTypes];
+  } else if (pnpTypes.length) {
+    return pnpTypes;
+  }
+}
 function getOriginalAndResolvedFileName(fileName, host, traceEnabled) {
   const resolvedFileName = realPath(fileName, host, traceEnabled);
   const pathsAreEqual = arePathsEqual(fileName, resolvedFileName, host);
@@ -38864,7 +38923,15 @@ function loadModuleFromNearestNodeModulesDirectoryWorker(extensions, moduleName,
     return lookup(secondaryExtensions);
   }
   function lookup(extensions2) {
-    return forEachAncestorDirectory(normalizeSlashes(directory), (ancestorDirectory) => {
+    const issuer = normalizeSlashes(directory);
+    if (getPnpApi(issuer)) {
+      const resolutionFromCache = tryFindNonRelativeModuleNameInCache(cache, moduleName, mode, issuer, redirectedReference, state);
+      if (resolutionFromCache) {
+        return resolutionFromCache;
+      }
+      return toSearchResult(loadModuleFromImmediateNodeModulesDirectoryPnP(extensions2, moduleName, issuer, state, typesScopeOnly, cache, redirectedReference));
+    }
+    return forEachAncestorDirectory(issuer, (ancestorDirectory) => {
       if (getBaseFileName(ancestorDirectory) !== "node_modules") {
         const resolutionFromCache = tryFindNonRelativeModuleNameInCache(cache, moduleName, mode, ancestorDirectory, redirectedReference, state);
         if (resolutionFromCache) {
@@ -38899,11 +38966,40 @@ function loadModuleFromImmediateNodeModulesDirectory(extensions, moduleName, dir
     return loadModuleFromSpecificNodeModulesDirectory(4 /* Declaration */, mangleScopedPackageNameWithTrace(moduleName, state), nodeModulesAtTypes2, nodeModulesAtTypesExists, state, cache, redirectedReference);
   }
 }
+function loadModuleFromImmediateNodeModulesDirectoryPnP(extensions, moduleName, directory, state, typesScopeOnly, cache, redirectedReference) {
+  const issuer = normalizeSlashes(directory);
+  if (!typesScopeOnly) {
+    const packageResult = tryLoadModuleUsingPnpResolution(extensions, moduleName, issuer, state, cache, redirectedReference);
+    if (packageResult) {
+      return packageResult;
+    }
+  }
+  if (extensions & 4 /* Declaration */) {
+    return tryLoadModuleUsingPnpResolution(4 /* Declaration */, `@types/${mangleScopedPackageNameWithTrace(moduleName, state)}`, issuer, state, cache, redirectedReference);
+  }
+}
 function loadModuleFromSpecificNodeModulesDirectory(extensions, moduleName, nodeModulesDirectory, nodeModulesDirectoryExists, state, cache, redirectedReference) {
-  var _a, _b;
   const candidate = normalizePath(combinePaths(nodeModulesDirectory, moduleName));
   const { packageName, rest } = parsePackageName(moduleName);
   const packageDirectory = combinePaths(nodeModulesDirectory, packageName);
+  return loadModuleFromSpecificNodeModulesDirectoryImpl(extensions, nodeModulesDirectoryExists, state, cache, redirectedReference, candidate, rest, packageDirectory);
+}
+function loadModuleFromPnpResolution(extensions, packageDirectory, rest, state, cache, redirectedReference) {
+  const candidate = normalizePath(combinePaths(packageDirectory, rest));
+  return loadModuleFromSpecificNodeModulesDirectoryImpl(
+    extensions,
+    /*nodeModulesDirectoryExists*/
+    true,
+    state,
+    cache,
+    redirectedReference,
+    candidate,
+    rest,
+    packageDirectory
+  );
+}
+function loadModuleFromSpecificNodeModulesDirectoryImpl(extensions, nodeModulesDirectoryExists, state, cache, redirectedReference, candidate, rest, packageDirectory) {
+  var _a, _b;
   let rootPackageInfo;
   let packageInfo = getPackageJsonInfo(candidate, !nodeModulesDirectoryExists, state);
   if (rest !== "" && packageInfo && (!(state.features & 8 /* Exports */) || !hasProperty(((_a = rootPackageInfo = getPackageJsonInfo(packageDirectory, !nodeModulesDirectoryExists, state)) == null ? void 0 : _a.contents.packageJsonContent) ?? emptyArray, "exports"))) {
@@ -39200,6 +39296,26 @@ function traceIfEnabled(state, diagnostic, ...args) {
 function useCaseSensitiveFileNames(state) {
   return !state.host.useCaseSensitiveFileNames ? true : typeof state.host.useCaseSensitiveFileNames === "boolean" ? state.host.useCaseSensitiveFileNames : state.host.useCaseSensitiveFileNames();
 }
+var { isBuiltin } = require("module");
+var failedLookups = /* @__PURE__ */ new Set();
+function loadPnpPackageResolution(packageName, containingDirectory) {
+  if (failedLookups.has(packageName)) {
+    return;
+  }
+  try {
+    const resolution = getPnpApi(containingDirectory).resolveToUnqualified(packageName, `${containingDirectory}/`, { considerBuiltins: false });
+    return normalizeSlashes(resolution).replace(/\/$/, "");
+  } catch {
+    if (isBuiltin(packageName) || packageName.startsWith("@types/") && isBuiltin(packageName.slice(7)) || (packageName.startsWith("@typescript/") || packageName.startsWith("@types/typescript__"))) {
+      failedLookups.add(packageName);
+    }
+  }
+}
+function tryLoadModuleUsingPnpResolution(extensions, moduleName, containingDirectory, state, cache, redirectedReference) {
+  const { packageName, rest } = parsePackageName(moduleName);
+  const packageResolution = loadPnpPackageResolution(packageName, containingDirectory);
+  return packageResolution ? loadModuleFromPnpResolution(extensions, packageResolution, rest, state, cache, redirectedReference) : void 0;
+}

 // src/compiler/binder.ts
 function getModuleInstanceState(node, visited) {
@@ -42494,7 +42610,15 @@ function getAllModulePathsWorker(importingFileName, importedFileName, host) {
     /*preferSymlinks*/
     true,
     (path, isRedirect) => {
-      const isInNodeModules = pathContainsNodeModules(path);
+      let isInNodeModules = pathContainsNodeModules(path);
+      const pnpapi = getPnpApi(path);
+      if (!isInNodeModules && pnpapi) {
+        const fromLocator = pnpapi.findPackageLocator(importingFileName);
+        const toLocator = pnpapi.findPackageLocator(path);
+        if (fromLocator && toLocator && fromLocator !== toLocator) {
+          isInNodeModules = true;
+        }
+      }
       allFileNames.set(path, { path: getCanonicalFileName(path), isRedirect, isInNodeModules });
       importedFileFromNodeModules = importedFileFromNodeModules || isInNodeModules;
     }
@@ -42685,7 +42809,41 @@ function tryGetModuleNameAsNodeModule({ path, isRedirect }, { getCanonicalFileNa
   if (!host.fileExists || !host.readFile) {
     return void 0;
   }
-  const parts = getNodeModulePathParts(path);
+  let parts = getNodeModulePathParts(path);
+  let pnpPackageName;
+  const pnpApi = getPnpApi(path);
+  if (pnpApi) {
+    const fromLocator = pnpApi.findPackageLocator(importingSourceFile.fileName);
+    const toLocator = pnpApi.findPackageLocator(path);
+    if (fromLocator === toLocator) {
+      return void 0;
+    }
+    if (fromLocator && toLocator) {
+      const fromInfo = pnpApi.getPackageInformation(fromLocator);
+      if (toLocator.reference === fromInfo.packageDependencies.get(toLocator.name)) {
+        pnpPackageName = toLocator.name;
+      } else {
+        for (const [name, reference] of fromInfo.packageDependencies) {
+          if (Array.isArray(reference)) {
+            if (reference[0] === toLocator.name && reference[1] === toLocator.reference) {
+              pnpPackageName = name;
+              break;
+            }
+          }
+        }
+      }
+      if (!parts) {
+        const toInfo = pnpApi.getPackageInformation(toLocator);
+        parts = {
+          topLevelNodeModulesIndex: void 0,
+          topLevelPackageNameIndex: void 0,
+          // The last character from packageLocation is the trailing "/", we want to point to it
+          packageRootIndex: toInfo.packageLocation.length - 1,
+          fileNameIndex: path.lastIndexOf(`/`)
+        };
+      }
+    }
+  }
   if (!parts) {
     return void 0;
   }
@@ -42723,14 +42881,16 @@ function tryGetModuleNameAsNodeModule({ path, isRedirect }, { getCanonicalFileNa
   if (isRedirect && !isPackageRootPath) {
     return void 0;
   }
-  const globalTypingsCacheLocation = host.getGlobalTypingsCacheLocation && host.getGlobalTypingsCacheLocation();
-  const pathToTopLevelNodeModules = getCanonicalFileName(moduleSpecifier.substring(0, parts.topLevelNodeModulesIndex));
-  if (!(startsWith(sourceDirectory, pathToTopLevelNodeModules) || globalTypingsCacheLocation && startsWith(getCanonicalFileName(globalTypingsCacheLocation), pathToTopLevelNodeModules))) {
-    return void 0;
+  if (typeof process.versions.pnp === "undefined") {
+    const globalTypingsCacheLocation = host.getGlobalTypingsCacheLocation && host.getGlobalTypingsCacheLocation();
+    const pathToTopLevelNodeModules = getCanonicalFileName(moduleSpecifier.substring(0, parts.topLevelNodeModulesIndex));
+    if (!(startsWith(sourceDirectory, pathToTopLevelNodeModules) || globalTypingsCacheLocation && startsWith(getCanonicalFileName(globalTypingsCacheLocation), pathToTopLevelNodeModules))) {
+      return void 0;
+    }
   }
-  const nodeModulesDirectoryName = moduleSpecifier.substring(parts.topLevelPackageNameIndex + 1);
-  const packageName = getPackageNameFromTypesPackageName(nodeModulesDirectoryName);
-  return getEmitModuleResolutionKind(options) === 1 /* Classic */ && packageName === nodeModulesDirectoryName ? void 0 : packageName;
+  const nodeModulesDirectoryName = typeof pnpPackageName !== "undefined" ? pnpPackageName + moduleSpecifier.substring(parts.packageRootIndex) : moduleSpecifier.substring(parts.topLevelPackageNameIndex + 1);
+  const packageNameFromPath = getPackageNameFromTypesPackageName(nodeModulesDirectoryName);
+  return getEmitModuleResolutionKind(options) === 1 /* Classic */ && packageNameFromPath === nodeModulesDirectoryName ? void 0 : packageNameFromPath;
   function tryDirectoryWithPackageJson(packageRootIndex) {
     var _a, _b;
     const packageRootPath = path.substring(0, packageRootIndex);
@@ -42743,9 +42903,9 @@ function tryGetModuleNameAsNodeModule({ path, isRedirect }, { getCanonicalFileNa
       const importMode = overrideMode || importingSourceFile.impliedNodeFormat;
       if (getResolvePackageJsonExports(options)) {
         const nodeModulesDirectoryName2 = packageRootPath.substring(parts.topLevelPackageNameIndex + 1);
-        const packageName2 = getPackageNameFromTypesPackageName(nodeModulesDirectoryName2);
+        const packageName = getPackageNameFromTypesPackageName(pnpPackageName ? pnpPackageName : nodeModulesDirectoryName2);
         const conditions = getConditions(options, importMode === 99 /* ESNext */);
-        const fromExports = packageJsonContent.exports ? tryGetModuleNameFromExports(options, path, packageRootPath, packageName2, packageJsonContent.exports, conditions) : void 0;
+        const fromExports = packageJsonContent.exports ? tryGetModuleNameFromExports(options, path, packageRootPath, packageName, packageJsonContent.exports, conditions) : void 0;
         if (fromExports) {
           const withJsExtension = !hasTSFileExtension(fromExports.moduleFileToTry) ? fromExports : { moduleFileToTry: removeFileExtension(fromExports.moduleFileToTry) + tryGetJSExtensionForFile(fromExports.moduleFileToTry, options) };
           return { ...withJsExtension, verbatimFromExports: true };
@@ -45473,7 +45633,7 @@ function createTypeChecker(host) {
     }
   }
   function getTargetofModuleDefault(moduleSymbol, node, dontResolveAlias) {
-    var _a;
+    var _a, _b;
     let exportDefaultSymbol;
     if (isShorthandAmbientModuleSymbol(moduleSymbol)) {
       exportDefaultSymbol = moduleSymbol;
@@ -45482,7 +45642,7 @@ function createTypeChecker(host) {
     }
     const file = (_a = moduleSymbol.declarations) == null ? void 0 : _a.find(isSourceFile);
     const specifier = getModuleSpecifierForImportOrExport(node);
-    if (!specifier) {
+    if (!specifier || moduleSymbol.flags === 512 && ((_b = moduleSymbol.exports) == null ? void 0 : _b.has("default"))) {
       return exportDefaultSymbol;
     }
     const hasDefaultOnly = isOnlyImportedAsDefault(specifier);
@@ -108978,10 +109138,11 @@ function getOutputExtension(fileName, options) {
   return fileExtensionIs(fileName, ".json" /* Json */) ? ".json" /* Json */ : options.jsx === 1 /* Preserve */ && fileExtensionIsOneOf(fileName, [".jsx" /* Jsx */, ".tsx" /* Tsx */]) ? ".jsx" /* Jsx */ : fileExtensionIsOneOf(fileName, [".mts" /* Mts */, ".mjs" /* Mjs */]) ? ".mjs" /* Mjs */ : fileExtensionIsOneOf(fileName, [".cts" /* Cts */, ".cjs" /* Cjs */]) ? ".cjs" /* Cjs */ : ".js" /* Js */;
 }
 function getOutputPathWithoutChangingExt(inputFileName, configFile, ignoreCase, outputDir, getCommonSourceDirectory2) {
-  return outputDir ? resolvePath(
-    outputDir,
-    getRelativePathFromDirectory(getCommonSourceDirectory2 ? getCommonSourceDirectory2() : getCommonSourceDirectoryOfConfig(configFile, ignoreCase), inputFileName, ignoreCase)
-  ) : inputFileName;
+  if (outputDir) {
+    const commonDir = getCommonSourceDirectory2 ? getCommonSourceDirectory2() : getCommonSourceDirectoryOfConfig(configFile, ignoreCase);
+    return `${outputDir}${directorySeparator}${inputFileName.slice(commonDir.length)}`;
+  }
+  return inputFileName;
 }
 function getOutputDeclarationFileName(inputFileName, configFile, ignoreCase, getCommonSourceDirectory2) {
   return changeExtension(
@@ -122349,6 +122510,11 @@ function createWatchProgram(host) {
   if (configFileName) {
     configFileWatcher = watchFile2(configFileName, scheduleProgramReload, 2e3 /* High */, watchOptions, WatchType.ConfigFile);
   }
+  let pnpFileWatcher;
+  const pnpApiPath = getPnpApiPath(__filename);
+  if (pnpApiPath) {
+    pnpFileWatcher = watchFile2(pnpApiPath, scheduleResolutionReload, 2e3 /* High */, watchOptions, WatchType.ConfigFile);
+  }
   const compilerHost = createCompilerHostFromProgramHost(host, () => compilerOptions, directoryStructureHost);
   setGetSourceFileAsHashVersioned(compilerHost);
   const getNewSourceFile = compilerHost.getSourceFile;
@@ -122412,6 +122578,10 @@ function createWatchProgram(host) {
       configFileWatcher.close();
       configFileWatcher = void 0;
     }
+    if (pnpFileWatcher) {
+      pnpFileWatcher.close();
+      pnpFileWatcher = void 0;
+    }
     extendedConfigCache == null ? void 0 : extendedConfigCache.clear();
     extendedConfigCache = void 0;
     if (sharedExtendedConfigFileWatchers) {
@@ -122447,7 +122617,7 @@ function createWatchProgram(host) {
   function getCurrentProgram() {
     return builderProgram && builderProgram.getProgramOrUndefined();
   }
-  function synchronizeProgram() {
+  function synchronizeProgram(forceAllFilesAsInvalidated = false) {
     writeLog(`Synchronizing program`);
     Debug.assert(compilerOptions);
     Debug.assert(rootFileNames);
@@ -122459,7 +122629,7 @@ function createWatchProgram(host) {
         resolutionCache.onChangesAffectModuleResolution();
       }
     }
-    const { hasInvalidatedResolutions, hasInvalidatedLibResolutions } = resolutionCache.createHasInvalidatedResolutions(customHasInvalidatedResolutions, customHasInvalidLibResolutions);
+    const { hasInvalidatedResolutions, hasInvalidatedLibResolutions } = resolutionCache.createHasInvalidatedResolutions(forceAllFilesAsInvalidated ? returnTrue : customHasInvalidatedResolutions, customHasInvalidLibResolutions);
     const {
       originalReadFile,
       originalFileExists,
@@ -122667,13 +122837,19 @@ function createWatchProgram(host) {
     reloadLevel = 2 /* Full */;
     scheduleProgramUpdate();
   }
+  function scheduleResolutionReload() {
+    writeLog("Clearing resolutions");
+    resolutionCache.clear();
+    reloadLevel = 3 /* Resolutions */;
+    scheduleProgramUpdate();
+  }
   function updateProgramWithWatchStatus() {
     timerToUpdateProgram = void 0;
     reportFileChangeDetectedOnCreateProgram = true;
     updateProgram();
   }
   function updateProgram() {
-    var _a, _b, _c, _d;
+    var _a, _b, _c, _d, _e;
     switch (reloadLevel) {
       case 1 /* Partial */:
         (_a = perfLogger) == null ? void 0 : _a.logStartUpdateProgram("PartialConfigReload");
@@ -122683,12 +122859,19 @@ function createWatchProgram(host) {
         (_b = perfLogger) == null ? void 0 : _b.logStartUpdateProgram("FullConfigReload");
         reloadConfigFile();
         break;
+      case 3 /* Resolutions */:
+        (_c = perfLogger) == null ? void 0 : _c.logStartUpdateProgram("SynchronizeProgramWithResolutions");
+        synchronizeProgram(
+          /*forceAllFilesAsInvalidated*/
+          true
+        );
+        break;
       default:
-        (_c = perfLogger) == null ? void 0 : _c.logStartUpdateProgram("SynchronizeProgram");
+        (_d = perfLogger) == null ? void 0 : _d.logStartUpdateProgram("SynchronizeProgram");
         synchronizeProgram();
         break;
     }
-    (_d = perfLogger) == null ? void 0 : _d.logStopUpdateProgram("Done");
+    (_e = perfLogger) == null ? void 0 : _e.logStopUpdateProgram("Done");
     return getCurrentBuilderProgram();
   }
   function reloadFileNamesFromConfigFile() {
diff --git a/lib/tsserver.js b/lib/tsserver.js
index e8341e70e..be14d713e 100644
semver exclusivity >=5.3.0
--- a/lib/tsserver.js
+++ b/lib/tsserver.js
@@ -8560,6 +8560,9 @@ var sys = (() => {
       return Buffer2.from && Buffer2.from !== Int8Array.from ? Buffer2.from(input, encoding) : new Buffer2(input, encoding);
     }
     function isFileSystemCaseSensitive() {
+      if (typeof process.versions.pnp !== `undefined`) {
+        return true;
+      }
       if (platform === "win32" || platform === "win64") {
         return false;
       }
@@ -8947,12 +8950,19 @@ function getAnyExtensionFromPath(path, extensions, ignoreCase) {
   }
   return "";
 }
+var pathComponentsCache = /* @__PURE__ */ new Map();
 function pathComponents(path, rootLength) {
+  const cached = pathComponentsCache.get(path);
+  if (cached) {
+    return cached.slice();
+  }
   const root = path.substring(0, rootLength);
   const rest = path.substring(rootLength).split(directorySeparator);
   if (rest.length && !lastOrUndefined(rest))
     rest.pop();
-  return [root, ...rest];
+  const res = [root, ...rest];
+  pathComponentsCache.set(path, res);
+  return res;
 }
 function getPathComponents(path, currentDirectory = "") {
   path = combinePaths(currentDirectory, path);
@@ -41251,6 +41261,67 @@ function getDefaultValueForOption(option) {
   }
 }

+// src/compiler/pnp.ts
+function getPnpApi(path) {
+  if (typeof process.versions.pnp === "undefined") {
+    return;
+  }
+  const { findPnpApi } = require("module");
+  if (findPnpApi) {
+    return findPnpApi(`${path}/`);
+  }
+}
+function getPnpApiPath(path) {
+  var _a;
+  return (_a = getPnpApi(path)) == null ? void 0 : _a.resolveRequest(
+    "pnpapi",
+    /*issuer*/
+    null
+  );
+}
+function getPnpTypeRoots(currentDirectory) {
+  const pnpApi = getPnpApi(currentDirectory);
+  if (!pnpApi) {
+    return [];
+  }
+  currentDirectory = resolvePath(currentDirectory);
+  const currentPackage = pnpApi.findPackageLocator(`${currentDirectory}/`);
+  if (!currentPackage) {
+    return [];
+  }
+  const { packageDependencies } = pnpApi.getPackageInformation(currentPackage);
+  const typeRoots = [];
+  for (const [name, referencish] of Array.from(packageDependencies.entries())) {
+    if (name.startsWith(`@types/`) && referencish !== null) {
+      const dependencyLocator = pnpApi.getLocator(name, referencish);
+      const { packageLocation } = pnpApi.getPackageInformation(dependencyLocator);
+      typeRoots.push(getDirectoryPath(packageLocation));
+    }
+  }
+  return typeRoots;
+}
+function isImportablePathPnp(fromPath, toPath3) {
+  const pnpApi = getPnpApi(fromPath);
+  const fromLocator = pnpApi.findPackageLocator(fromPath);
+  const toLocator = pnpApi.findPackageLocator(toPath3);
+  if (toLocator === null) {
+    return false;
+  }
+  const fromInfo = pnpApi.getPackageInformation(fromLocator);
+  const toReference = fromInfo.packageDependencies.get(toLocator.name);
+  if (toReference) {
+    return toReference === toLocator.reference;
+  }
+  for (const reference of fromInfo.packageDependencies.values()) {
+    if (Array.isArray(reference)) {
+      if (reference[0] === toLocator.name && reference[1] === toLocator.reference) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
 // src/compiler/moduleNameResolver.ts
 function trace(host, message, ...args) {
   host.trace(formatMessage(message, ...args));
@@ -41473,7 +41544,7 @@ function getEffectiveTypeRoots(options, host) {
     return getDefaultTypeRoots(currentDirectory);
   }
 }
-function getDefaultTypeRoots(currentDirectory) {
+function getNodeModulesTypeRoots(currentDirectory) {
   let typeRoots;
   forEachAncestorDirectory(normalizePath(currentDirectory), (directory) => {
     const atTypes = combinePaths(directory, nodeModulesAtTypes);
@@ -41486,6 +41557,15 @@ function arePathsEqual(path1, path2, host) {
   const useCaseSensitiveFileNames2 = typeof host.useCaseSensitiveFileNames === "function" ? host.useCaseSensitiveFileNames() : host.useCaseSensitiveFileNames;
   return comparePaths(path1, path2, !useCaseSensitiveFileNames2) === 0 /* EqualTo */;
 }
+function getDefaultTypeRoots(currentDirectory) {
+  const nmTypes = getNodeModulesTypeRoots(currentDirectory);
+  const pnpTypes = getPnpTypeRoots(currentDirectory);
+  if (nmTypes == null ? void 0 : nmTypes.length) {
+    return [...nmTypes, ...pnpTypes];
+  } else if (pnpTypes.length) {
+    return pnpTypes;
+  }
+}
 function getOriginalAndResolvedFileName(fileName, host, traceEnabled) {
   const resolvedFileName = realPath(fileName, host, traceEnabled);
   const pathsAreEqual = arePathsEqual(fileName, resolvedFileName, host);
@@ -41725,6 +41805,21 @@ function getConditions(options, esmMode) {
 }
 function resolvePackageNameToPackageJson(packageName, containingDirectory, options, host, cache) {
   const moduleResolutionState = getTemporaryModuleResolutionState(cache == null ? void 0 : cache.getPackageJsonInfoCache(), host, options);
+  const pnpapi = getPnpApi(containingDirectory);
+  if (pnpapi) {
+    try {
+      const resolution = pnpapi.resolveToUnqualified(packageName, `${containingDirectory}/`, { considerBuiltins: false });
+      const candidate = normalizeSlashes(resolution).replace(/\/$/, "");
+      return getPackageJsonInfo(
+        candidate,
+        /*onlyRecordFailures*/
+        false,
+        moduleResolutionState
+      );
+    } catch {
+      return;
+    }
+  }
   return forEachAncestorDirectory(containingDirectory, (ancestorDirectory) => {
     if (getBaseFileName(ancestorDirectory) !== "node_modules") {
       const nodeModulesFolder = combinePaths(ancestorDirectory, "node_modules");
@@ -43505,7 +43600,15 @@ function loadModuleFromNearestNodeModulesDirectoryWorker(extensions, moduleName,
     return lookup(secondaryExtensions);
   }
   function lookup(extensions2) {
-    return forEachAncestorDirectory(normalizeSlashes(directory), (ancestorDirectory) => {
+    const issuer = normalizeSlashes(directory);
+    if (getPnpApi(issuer)) {
+      const resolutionFromCache = tryFindNonRelativeModuleNameInCache(cache, moduleName, mode, issuer, redirectedReference, state);
+      if (resolutionFromCache) {
+        return resolutionFromCache;
+      }
+      return toSearchResult(loadModuleFromImmediateNodeModulesDirectoryPnP(extensions2, moduleName, issuer, state, typesScopeOnly, cache, redirectedReference));
+    }
+    return forEachAncestorDirectory(issuer, (ancestorDirectory) => {
       if (getBaseFileName(ancestorDirectory) !== "node_modules") {
         const resolutionFromCache = tryFindNonRelativeModuleNameInCache(cache, moduleName, mode, ancestorDirectory, redirectedReference, state);
         if (resolutionFromCache) {
@@ -43540,11 +43643,40 @@ function loadModuleFromImmediateNodeModulesDirectory(extensions, moduleName, dir
     return loadModuleFromSpecificNodeModulesDirectory(4 /* Declaration */, mangleScopedPackageNameWithTrace(moduleName, state), nodeModulesAtTypes2, nodeModulesAtTypesExists, state, cache, redirectedReference);
   }
 }
+function loadModuleFromImmediateNodeModulesDirectoryPnP(extensions, moduleName, directory, state, typesScopeOnly, cache, redirectedReference) {
+  const issuer = normalizeSlashes(directory);
+  if (!typesScopeOnly) {
+    const packageResult = tryLoadModuleUsingPnpResolution(extensions, moduleName, issuer, state, cache, redirectedReference);
+    if (packageResult) {
+      return packageResult;
+    }
+  }
+  if (extensions & 4 /* Declaration */) {
+    return tryLoadModuleUsingPnpResolution(4 /* Declaration */, `@types/${mangleScopedPackageNameWithTrace(moduleName, state)}`, issuer, state, cache, redirectedReference);
+  }
+}
 function loadModuleFromSpecificNodeModulesDirectory(extensions, moduleName, nodeModulesDirectory, nodeModulesDirectoryExists, state, cache, redirectedReference) {
-  var _a, _b;
   const candidate = normalizePath(combinePaths(nodeModulesDirectory, moduleName));
   const { packageName, rest } = parsePackageName(moduleName);
   const packageDirectory = combinePaths(nodeModulesDirectory, packageName);
+  return loadModuleFromSpecificNodeModulesDirectoryImpl(extensions, nodeModulesDirectoryExists, state, cache, redirectedReference, candidate, rest, packageDirectory);
+}
+function loadModuleFromPnpResolution(extensions, packageDirectory, rest, state, cache, redirectedReference) {
+  const candidate = normalizePath(combinePaths(packageDirectory, rest));
+  return loadModuleFromSpecificNodeModulesDirectoryImpl(
+    extensions,
+    /*nodeModulesDirectoryExists*/
+    true,
+    state,
+    cache,
+    redirectedReference,
+    candidate,
+    rest,
+    packageDirectory
+  );
+}
+function loadModuleFromSpecificNodeModulesDirectoryImpl(extensions, nodeModulesDirectoryExists, state, cache, redirectedReference, candidate, rest, packageDirectory) {
+  var _a, _b;
   let rootPackageInfo;
   let packageInfo = getPackageJsonInfo(candidate, !nodeModulesDirectoryExists, state);
   if (rest !== "" && packageInfo && (!(state.features & 8 /* Exports */) || !hasProperty(((_a = rootPackageInfo = getPackageJsonInfo(packageDirectory, !nodeModulesDirectoryExists, state)) == null ? void 0 : _a.contents.packageJsonContent) ?? emptyArray, "exports"))) {
@@ -43841,6 +43973,26 @@ function traceIfEnabled(state, diagnostic, ...args) {
 function useCaseSensitiveFileNames(state) {
   return !state.host.useCaseSensitiveFileNames ? true : typeof state.host.useCaseSensitiveFileNames === "boolean" ? state.host.useCaseSensitiveFileNames : state.host.useCaseSensitiveFileNames();
 }
+var { isBuiltin } = require("module");
+var failedLookups = /* @__PURE__ */ new Set();
+function loadPnpPackageResolution(packageName, containingDirectory) {
+  if (failedLookups.has(packageName)) {
+    return;
+  }
+  try {
+    const resolution = getPnpApi(containingDirectory).resolveToUnqualified(packageName, `${containingDirectory}/`, { considerBuiltins: false });
+    return normalizeSlashes(resolution).replace(/\/$/, "");
+  } catch {
+    if (isBuiltin(packageName) || packageName.startsWith("@types/") && isBuiltin(packageName.slice(7)) || (packageName.startsWith("@typescript/") || packageName.startsWith("@types/typescript__"))) {
+      failedLookups.add(packageName);
+    }
+  }
+}
+function tryLoadModuleUsingPnpResolution(extensions, moduleName, containingDirectory, state, cache, redirectedReference) {
+  const { packageName, rest } = parsePackageName(moduleName);
+  const packageResolution = loadPnpPackageResolution(packageName, containingDirectory);
+  return packageResolution ? loadModuleFromPnpResolution(extensions, packageResolution, rest, state, cache, redirectedReference) : void 0;
+}

 // src/compiler/binder.ts
 var ModuleInstanceState = /* @__PURE__ */ ((ModuleInstanceState2) => {
@@ -47198,7 +47350,15 @@ function getAllModulePathsWorker(importingFileName, importedFileName, host) {
     /*preferSymlinks*/
     true,
     (path, isRedirect) => {
-      const isInNodeModules = pathContainsNodeModules(path);
+      let isInNodeModules = pathContainsNodeModules(path);
+      const pnpapi = getPnpApi(path);
+      if (!isInNodeModules && pnpapi) {
+        const fromLocator = pnpapi.findPackageLocator(importingFileName);
+        const toLocator = pnpapi.findPackageLocator(path);
+        if (fromLocator && toLocator && fromLocator !== toLocator) {
+          isInNodeModules = true;
+        }
+      }
       allFileNames.set(path, { path: getCanonicalFileName(path), isRedirect, isInNodeModules });
       importedFileFromNodeModules = importedFileFromNodeModules || isInNodeModules;
     }
@@ -47389,7 +47549,41 @@ function tryGetModuleNameAsNodeModule({ path, isRedirect }, { getCanonicalFileNa
   if (!host.fileExists || !host.readFile) {
     return void 0;
   }
-  const parts = getNodeModulePathParts(path);
+  let parts = getNodeModulePathParts(path);
+  let pnpPackageName;
+  const pnpApi = getPnpApi(path);
+  if (pnpApi) {
+    const fromLocator = pnpApi.findPackageLocator(importingSourceFile.fileName);
+    const toLocator = pnpApi.findPackageLocator(path);
+    if (fromLocator === toLocator) {
+      return void 0;
+    }
+    if (fromLocator && toLocator) {
+      const fromInfo = pnpApi.getPackageInformation(fromLocator);
+      if (toLocator.reference === fromInfo.packageDependencies.get(toLocator.name)) {
+        pnpPackageName = toLocator.name;
+      } else {
+        for (const [name, reference] of fromInfo.packageDependencies) {
+          if (Array.isArray(reference)) {
+            if (reference[0] === toLocator.name && reference[1] === toLocator.reference) {
+              pnpPackageName = name;
+              break;
+            }
+          }
+        }
+      }
+      if (!parts) {
+        const toInfo = pnpApi.getPackageInformation(toLocator);
+        parts = {
+          topLevelNodeModulesIndex: void 0,
+          topLevelPackageNameIndex: void 0,
+          // The last character from packageLocation is the trailing "/", we want to point to it
+          packageRootIndex: toInfo.packageLocation.length - 1,
+          fileNameIndex: path.lastIndexOf(`/`)
+        };
+      }
+    }
+  }
   if (!parts) {
     return void 0;
   }
@@ -47427,14 +47621,16 @@ function tryGetModuleNameAsNodeModule({ path, isRedirect }, { getCanonicalFileNa
   if (isRedirect && !isPackageRootPath) {
     return void 0;
   }
-  const globalTypingsCacheLocation = host.getGlobalTypingsCacheLocation && host.getGlobalTypingsCacheLocation();
-  const pathToTopLevelNodeModules = getCanonicalFileName(moduleSpecifier.substring(0, parts.topLevelNodeModulesIndex));
-  if (!(startsWith(sourceDirectory, pathToTopLevelNodeModules) || globalTypingsCacheLocation && startsWith(getCanonicalFileName(globalTypingsCacheLocation), pathToTopLevelNodeModules))) {
-    return void 0;
+  if (typeof process.versions.pnp === "undefined") {
+    const globalTypingsCacheLocation = host.getGlobalTypingsCacheLocation && host.getGlobalTypingsCacheLocation();
+    const pathToTopLevelNodeModules = getCanonicalFileName(moduleSpecifier.substring(0, parts.topLevelNodeModulesIndex));
+    if (!(startsWith(sourceDirectory, pathToTopLevelNodeModules) || globalTypingsCacheLocation && startsWith(getCanonicalFileName(globalTypingsCacheLocation), pathToTopLevelNodeModules))) {
+      return void 0;
+    }
   }
-  const nodeModulesDirectoryName = moduleSpecifier.substring(parts.topLevelPackageNameIndex + 1);
-  const packageName = getPackageNameFromTypesPackageName(nodeModulesDirectoryName);
-  return getEmitModuleResolutionKind(options) === 1 /* Classic */ && packageName === nodeModulesDirectoryName ? void 0 : packageName;
+  const nodeModulesDirectoryName = typeof pnpPackageName !== "undefined" ? pnpPackageName + moduleSpecifier.substring(parts.packageRootIndex) : moduleSpecifier.substring(parts.topLevelPackageNameIndex + 1);
+  const packageNameFromPath = getPackageNameFromTypesPackageName(nodeModulesDirectoryName);
+  return getEmitModuleResolutionKind(options) === 1 /* Classic */ && packageNameFromPath === nodeModulesDirectoryName ? void 0 : packageNameFromPath;
   function tryDirectoryWithPackageJson(packageRootIndex) {
     var _a, _b;
     const packageRootPath = path.substring(0, packageRootIndex);
@@ -47447,9 +47643,9 @@ function tryGetModuleNameAsNodeModule({ path, isRedirect }, { getCanonicalFileNa
       const importMode = overrideMode || importingSourceFile.impliedNodeFormat;
       if (getResolvePackageJsonExports(options)) {
         const nodeModulesDirectoryName2 = packageRootPath.substring(parts.topLevelPackageNameIndex + 1);
-        const packageName2 = getPackageNameFromTypesPackageName(nodeModulesDirectoryName2);
+        const packageName = getPackageNameFromTypesPackageName(pnpPackageName ? pnpPackageName : nodeModulesDirectoryName2);
         const conditions = getConditions(options, importMode === 99 /* ESNext */);
-        const fromExports = packageJsonContent.exports ? tryGetModuleNameFromExports(options, path, packageRootPath, packageName2, packageJsonContent.exports, conditions) : void 0;
+        const fromExports = packageJsonContent.exports ? tryGetModuleNameFromExports(options, path, packageRootPath, packageName, packageJsonContent.exports, conditions) : void 0;
         if (fromExports) {
           const withJsExtension = !hasTSFileExtension(fromExports.moduleFileToTry) ? fromExports : { moduleFileToTry: removeFileExtension(fromExports.moduleFileToTry) + tryGetJSExtensionForFile(fromExports.moduleFileToTry, options) };
           return { ...withJsExtension, verbatimFromExports: true };
@@ -50177,7 +50373,7 @@ function createTypeChecker(host) {
     }
   }
   function getTargetofModuleDefault(moduleSymbol, node, dontResolveAlias) {
-    var _a;
+    var _a, _b;
     let exportDefaultSymbol;
     if (isShorthandAmbientModuleSymbol(moduleSymbol)) {
       exportDefaultSymbol = moduleSymbol;
@@ -50186,7 +50382,7 @@ function createTypeChecker(host) {
     }
     const file = (_a = moduleSymbol.declarations) == null ? void 0 : _a.find(isSourceFile);
     const specifier = getModuleSpecifierForImportOrExport(node);
-    if (!specifier) {
+    if (!specifier || moduleSymbol.flags === 512 && ((_b = moduleSymbol.exports) == null ? void 0 : _b.has("default"))) {
       return exportDefaultSymbol;
     }
     const hasDefaultOnly = isOnlyImportedAsDefault(specifier);
@@ -113853,10 +114049,11 @@ function getOutputExtension(fileName, options) {
   return fileExtensionIs(fileName, ".json" /* Json */) ? ".json" /* Json */ : options.jsx === 1 /* Preserve */ && fileExtensionIsOneOf(fileName, [".jsx" /* Jsx */, ".tsx" /* Tsx */]) ? ".jsx" /* Jsx */ : fileExtensionIsOneOf(fileName, [".mts" /* Mts */, ".mjs" /* Mjs */]) ? ".mjs" /* Mjs */ : fileExtensionIsOneOf(fileName, [".cts" /* Cts */, ".cjs" /* Cjs */]) ? ".cjs" /* Cjs */ : ".js" /* Js */;
 }
 function getOutputPathWithoutChangingExt(inputFileName, configFile, ignoreCase, outputDir, getCommonSourceDirectory2) {
-  return outputDir ? resolvePath(
-    outputDir,
-    getRelativePathFromDirectory(getCommonSourceDirectory2 ? getCommonSourceDirectory2() : getCommonSourceDirectoryOfConfig(configFile, ignoreCase), inputFileName, ignoreCase)
-  ) : inputFileName;
+  if (outputDir) {
+    const commonDir = getCommonSourceDirectory2 ? getCommonSourceDirectory2() : getCommonSourceDirectoryOfConfig(configFile, ignoreCase);
+    return `${outputDir}${directorySeparator}${inputFileName.slice(commonDir.length)}`;
+  }
+  return inputFileName;
 }
 function getOutputDeclarationFileName(inputFileName, configFile, ignoreCase, getCommonSourceDirectory2) {
   return changeExtension(
@@ -119473,6 +119670,7 @@ var ConfigFileProgramReloadLevel = /* @__PURE__ */ ((ConfigFileProgramReloadLeve
   ConfigFileProgramReloadLevel2[ConfigFileProgramReloadLevel2["None"] = 0] = "None";
   ConfigFileProgramReloadLevel2[ConfigFileProgramReloadLevel2["Partial"] = 1] = "Partial";
   ConfigFileProgramReloadLevel2[ConfigFileProgramReloadLevel2["Full"] = 2] = "Full";
+  ConfigFileProgramReloadLevel2[ConfigFileProgramReloadLevel2["Resolutions"] = 3] = "Resolutions";
   return ConfigFileProgramReloadLevel2;
 })(ConfigFileProgramReloadLevel || {});
 function updateSharedExtendedConfigFileWatcher(projectPath, options, extendedConfigFilesMap, createExtendedConfigFileWatch, toPath3) {
@@ -127349,6 +127547,11 @@ function createWatchProgram(host) {
   if (configFileName) {
     configFileWatcher = watchFile2(configFileName, scheduleProgramReload, 2e3 /* High */, watchOptions, WatchType.ConfigFile);
   }
+  let pnpFileWatcher;
+  const pnpApiPath = getPnpApiPath(__filename);
+  if (pnpApiPath) {
+    pnpFileWatcher = watchFile2(pnpApiPath, scheduleResolutionReload, 2e3 /* High */, watchOptions, WatchType.ConfigFile);
+  }
   const compilerHost = createCompilerHostFromProgramHost(host, () => compilerOptions, directoryStructureHost);
   setGetSourceFileAsHashVersioned(compilerHost);
   const getNewSourceFile = compilerHost.getSourceFile;
@@ -127412,6 +127615,10 @@ function createWatchProgram(host) {
       configFileWatcher.close();
       configFileWatcher = void 0;
     }
+    if (pnpFileWatcher) {
+      pnpFileWatcher.close();
+      pnpFileWatcher = void 0;
+    }
     extendedConfigCache == null ? void 0 : extendedConfigCache.clear();
     extendedConfigCache = void 0;
     if (sharedExtendedConfigFileWatchers) {
@@ -127447,7 +127654,7 @@ function createWatchProgram(host) {
   function getCurrentProgram() {
     return builderProgram && builderProgram.getProgramOrUndefined();
   }
-  function synchronizeProgram() {
+  function synchronizeProgram(forceAllFilesAsInvalidated = false) {
     writeLog(`Synchronizing program`);
     Debug.assert(compilerOptions);
     Debug.assert(rootFileNames);
@@ -127459,7 +127666,7 @@ function createWatchProgram(host) {
         resolutionCache.onChangesAffectModuleResolution();
       }
     }
-    const { hasInvalidatedResolutions, hasInvalidatedLibResolutions } = resolutionCache.createHasInvalidatedResolutions(customHasInvalidatedResolutions, customHasInvalidLibResolutions);
+    const { hasInvalidatedResolutions, hasInvalidatedLibResolutions } = resolutionCache.createHasInvalidatedResolutions(forceAllFilesAsInvalidated ? returnTrue : customHasInvalidatedResolutions, customHasInvalidLibResolutions);
     const {
       originalReadFile,
       originalFileExists,
@@ -127667,13 +127874,19 @@ function createWatchProgram(host) {
     reloadLevel = 2 /* Full */;
     scheduleProgramUpdate();
   }
+  function scheduleResolutionReload() {
+    writeLog("Clearing resolutions");
+    resolutionCache.clear();
+    reloadLevel = 3 /* Resolutions */;
+    scheduleProgramUpdate();
+  }
   function updateProgramWithWatchStatus() {
     timerToUpdateProgram = void 0;
     reportFileChangeDetectedOnCreateProgram = true;
     updateProgram();
   }
   function updateProgram() {
-    var _a, _b, _c, _d;
+    var _a, _b, _c, _d, _e;
     switch (reloadLevel) {
       case 1 /* Partial */:
         (_a = perfLogger) == null ? void 0 : _a.logStartUpdateProgram("PartialConfigReload");
@@ -127683,12 +127896,19 @@ function createWatchProgram(host) {
         (_b = perfLogger) == null ? void 0 : _b.logStartUpdateProgram("FullConfigReload");
         reloadConfigFile();
         break;
+      case 3 /* Resolutions */:
+        (_c = perfLogger) == null ? void 0 : _c.logStartUpdateProgram("SynchronizeProgramWithResolutions");
+        synchronizeProgram(
+          /*forceAllFilesAsInvalidated*/
+          true
+        );
+        break;
       default:
-        (_c = perfLogger) == null ? void 0 : _c.logStartUpdateProgram("SynchronizeProgram");
+        (_d = perfLogger) == null ? void 0 : _d.logStartUpdateProgram("SynchronizeProgram");
         synchronizeProgram();
         break;
     }
-    (_d = perfLogger) == null ? void 0 : _d.logStopUpdateProgram("Done");
+    (_e = perfLogger) == null ? void 0 : _e.logStopUpdateProgram("Done");
     return getCurrentBuilderProgram();
   }
   function reloadFileNamesFromConfigFile() {
@@ -133177,7 +133397,14 @@ function createPackageJsonImportFilter(fromFile, preferences, host) {
     return false;
   }
   function getNodeModulesPackageNameFromFileName(importedFileName, moduleSpecifierResolutionHost) {
-    if (!importedFileName.includes("node_modules")) {
+    const pnpapi = getPnpApi(importedFileName);
+    if (pnpapi) {
+      const fromLocator = pnpapi.findPackageLocator(fromFile.fileName);
+      const toLocator = pnpapi.findPackageLocator(importedFileName);
+      if (!(fromLocator && toLocator)) {
+        return void 0;
+      }
+    } else if (!importedFileName.includes("node_modules")) {
       return void 0;
     }
     const specifier = ts_moduleSpecifiers_exports.getNodeModulesPackageName(
@@ -133708,6 +133935,9 @@ function isImportableFile(program, from, to, preferences, packageJsonFilter, mod
   return hasImportablePath;
 }
 function isImportablePath(fromPath, toPath3, getCanonicalFileName, globalCachePath) {
+  if (getPnpApi(fromPath)) {
+    return isImportablePathPnp(fromPath, toPath3);
+  }
   const toNodeModules = forEachAncestorDirectory(toPath3, (ancestor) => getBaseFileName(ancestor) === "node_modules" ? ancestor : void 0);
   const toNodeModulesParent = toNodeModules && getDirectoryPath(getCanonicalFileName(toNodeModules));
   return toNodeModulesParent === void 0 || startsWith(getCanonicalFileName(fromPath), toNodeModulesParent) || !!globalCachePath && startsWith(getCanonicalFileName(globalCachePath), toNodeModulesParent);
@@ -161729,7 +161959,31 @@ function getCompletionEntriesForNonRelativeModules(fragment, scriptPath, mode, c
           );
         }
       };
-      if (fragmentDirectory && getResolvePackageJsonExports(compilerOptions)) {
+      const checkExports = (packageFile, packageDirectory, fragmentSubpath) => {
+        const packageJson = readJson(packageFile, host);
+        const exports = packageJson.exports;
+        if (exports) {
+          if (typeof exports !== "object" || exports === null) {
+            return true;
+          }
+          const keys = getOwnKeys(exports);
+          const conditions = getConditions(compilerOptions, mode === 99 /* ESNext */);
+          addCompletionEntriesFromPathsOrExports(
+            result,
+            fragmentSubpath,
+            packageDirectory,
+            extensionOptions,
+            host,
+            keys,
+            (key) => singleElementArray(getPatternFromFirstMatchingCondition(exports[key], conditions)),
+            comparePatternKeys
+          );
+          return true;
+        }
+        return false;
+      };
+      const shouldCheckExports = fragmentDirectory && getResolvePackageJsonExports(compilerOptions);
+      if (shouldCheckExports) {
         const nodeModulesDirectoryLookup = ancestorLookup;
         ancestorLookup = (ancestor) => {
           const components = getPathComponents(fragment);
@@ -161748,32 +162002,52 @@ function getCompletionEntriesForNonRelativeModules(fragment, scriptPath, mode, c
           const packageDirectory = combinePaths(ancestor, "node_modules", packagePath);
           const packageFile = combinePaths(packageDirectory, "package.json");
           if (tryFileExists(host, packageFile)) {
-            const packageJson = readJson(packageFile, host);
-            const exports = packageJson.exports;
-            if (exports) {
-              if (typeof exports !== "object" || exports === null) {
-                return;
+            const fragmentSubpath = components.join("/") + (components.length && hasTrailingDirectorySeparator(fragment) ? "/" : "");
+            if (checkExports(packageFile, packageDirectory, fragmentSubpath)) {
+              return;
+            }
+          }
+          return nodeModulesDirectoryLookup(ancestor);
+        };
+      }
+      const pnpApi = getPnpApi(scriptPath);
+      if (pnpApi) {
+        const pathRegExp = /^(?![a-zA-Z]:[\\/]|\\\\|\.{0,2}(?:\/|$))((?:@[^/]+\/)?[^/]+)\/*(.*|)$/;
+        const dependencyNameMatch = fragment.match(pathRegExp);
+        if (dependencyNameMatch) {
+          const [, dependencyName, subPath] = dependencyNameMatch;
+          let unqualified;
+          try {
+            unqualified = pnpApi.resolveToUnqualified(dependencyName, scriptPath, { considerBuiltins: false });
+          } catch {
+          }
+          if (unqualified) {
+            const packageDirectory = normalizePath(unqualified);
+            let shouldGetCompletions = true;
+            if (shouldCheckExports) {
+              const packageFile = combinePaths(packageDirectory, "package.json");
+              if (tryFileExists(host, packageFile) && checkExports(packageFile, packageDirectory, subPath)) {
+                shouldGetCompletions = false;
               }
-              const keys = getOwnKeys(exports);
-              const fragmentSubpath = components.join("/") + (components.length && hasTrailingDirectorySeparator(fragment) ? "/" : "");
-              const conditions = getConditions(compilerOptions, mode === 99 /* ESNext */);
-              addCompletionEntriesFromPathsOrExports(
-                result,
-                fragmentSubpath,
+            }
+            if (shouldGetCompletions) {
+              getCompletionEntriesForDirectoryFragment(
+                subPath,
                 packageDirectory,
                 extensionOptions,
                 host,
-                keys,
-                (key) => singleElementArray(getPatternFromFirstMatchingCondition(exports[key], conditions)),
-                comparePatternKeys
+                /*moduleSpecifierIsRelative*/
+                false,
+                /*exclude*/
+                void 0,
+                result
               );
-              return;
             }
           }
-          return nodeModulesDirectoryLookup(ancestor);
-        };
+        }
+      } else {
+        forEachAncestorDirectory(scriptPath, ancestorLookup);
       }
-      forEachAncestorDirectory(scriptPath, ancestorLookup);
     }
   }
   return arrayFrom(result.values());
@@ -161904,9 +162178,15 @@ function getCompletionEntriesFromTypings(host, options, scriptPath, fragmentDire
   for (const root of typeRoots) {
     getCompletionEntriesFromDirectories(root);
   }
-  for (const packageJson of findPackageJsons(scriptPath, host)) {
-    const typesDir = combinePaths(getDirectoryPath(packageJson), "node_modules/@types");
-    getCompletionEntriesFromDirectories(typesDir);
+  if (getPnpApi(scriptPath)) {
+    for (const root of getPnpTypeRoots(scriptPath)) {
+      getCompletionEntriesFromDirectories(root);
+    }
+  } else {
+    for (const packageJson of findPackageJsons(scriptPath, host)) {
+      const typesDir = combinePaths(getDirectoryPath(packageJson), "node_modules/@types");
+      getCompletionEntriesFromDirectories(typesDir);
+    }
   }
   return result;
   function getCompletionEntriesFromDirectories(directory) {
@@ -178712,6 +178992,29 @@ var ConfiguredProject2 = class extends Project3 {
     return this.projectReferences;
   }
   updateReferences(refs) {
+    if (typeof process.versions.pnp !== `undefined`) {
+      const basePath = this.getCurrentDirectory();
+      const getPnpPath = (path) => {
+        try {
+          const pnpApi = getPnpApi(`${path}/`);
+          if (!pnpApi) {
+            return path;
+          }
+          const targetLocator = pnpApi.findPackageLocator(`${path}/`);
+          const { packageLocation } = pnpApi.getPackageInformation(targetLocator);
+          const request = combinePaths(targetLocator.name, getRelativePathFromDirectory(
+            packageLocation,
+            path,
+            /*ignoreCase*/
+            false
+          ));
+          return pnpApi.resolveToUnqualified(request, `${basePath}/`);
+        } catch {
+          return path;
+        }
+      };
+      refs = refs == null ? void 0 : refs.map((r) => ({ ...r, path: getPnpPath(r.path) }));
+    }
     this.projectReferences = refs;
     this.potentialProjectReferences = void 0;
   }
@@ -179377,6 +179680,7 @@ var _ProjectService = class _ProjectService {
       log,
       getDetailWatchInfo
     );
+    this.pnpWatcher = this.watchPnpFile();
     (_a = opts.incrementalVerifier) == null ? void 0 : _a.call(opts, this);
   }
   toPath(fileName) {
@@ -181188,7 +181492,7 @@ Dynamic files must always be opened with service's current directory or service
     this.performanceEventHandler = performanceEventHandler;
   }
   setHostConfiguration(args) {
-    var _a;
+    var _a, _b;
     if (args.file) {
       const info = this.getScriptInfoForNormalizedPath(toNormalizedPath(args.file));
       if (info) {
@@ -181231,7 +181535,8 @@ Dynamic files must always be opened with service's current directory or service
       }
       if (args.watchOptions) {
         this.hostConfiguration.watchOptions = (_a = convertWatchOptions(args.watchOptions)) == null ? void 0 : _a.watchOptions;
-        this.logger.info(`Host watch options changed to ${JSON.stringify(this.hostConfiguration.watchOptions)}, it will be take effect for next watches.`);
+        (_b = this.pnpWatcher) == null ? void 0 : _b.close();
+        this.watchPnpFile();
       }
     }
   }
@@ -182178,6 +182483,29 @@ Dynamic files must always be opened with service's current directory or service
     });
   }
   /** @internal */
+  watchPnpFile() {
+    const pnpApiPath = getPnpApiPath(__filename);
+    if (!pnpApiPath) {
+      return;
+    }
+    return this.watchFactory.watchFile(
+      pnpApiPath,
+      () => {
+        this.forEachProject((project) => {
+          for (const info of project.getScriptInfos()) {
+            project.resolutionCache.invalidateResolutionOfFile(info.path);
+          }
+          project.markAsDirty();
+          updateProjectIfDirty(project);
+        });
+        this.delayEnsureProjectForOpenFiles();
+      },
+      250 /* Low */,
+      this.hostConfiguration.watchOptions,
+      WatchType.ConfigFile
+    );
+  }
+  /** @internal */
   watchPackageJsonFile(path) {
     const watchers = this.packageJsonFilesMap || (this.packageJsonFilesMap = /* @__PURE__ */ new Map());
     if (!watchers.has(path)) {
@@ -186766,6 +187094,10 @@ function initializeNodeSystem() {
         }
         try {
           const args = [combinePaths(libDirectory, "watchGuard.js"), path];
+          const pnpApiPath = getPnpApiPath(__filename);
+          if (pnpApiPath) {
+            args.unshift("-r", pnpApiPath);
+          }
           if (logger.hasLevel(3 /* verbose */)) {
             logger.info(`Starting ${process.execPath} with args:${stringifyIndented(args)}`);
           }
@@ -186971,6 +187303,10 @@ function startNodeSession(options, logger, cancellationToken) {
           break;
         }
       }
+      const pnpApiPath = getPnpApiPath(__filename);
+      if (pnpApiPath) {
+        execArgv.unshift("-r", pnpApiPath);
+      }
       const typingsInstaller = combinePaths(getDirectoryPath(sys.getExecutingFilePath()), "typingsInstaller.js");
       this.installer = childProcess.fork(typingsInstaller, args, { execArgv });
       this.installer.on("message", (m) => this.handleMessage(m));
diff --git a/lib/typescript.js b/lib/typescript.js
index 4c4d3228e..9fff08d16 100644
semver exclusivity >=5.3.0
--- a/lib/typescript.js
+++ b/lib/typescript.js
@@ -6332,6 +6332,9 @@ ${lanes.join("\n")}
             return Buffer2.from && Buffer2.from !== Int8Array.from ? Buffer2.from(input, encoding) : new Buffer2(input, encoding);
           }
           function isFileSystemCaseSensitive() {
+            if (typeof process.versions.pnp !== `undefined`) {
+              return true;
+            }
             if (platform === "win32" || platform === "win64") {
               return false;
             }
@@ -6715,11 +6718,17 @@ ${lanes.join("\n")}
     return "";
   }
   function pathComponents(path, rootLength) {
+    const cached = pathComponentsCache.get(path);
+    if (cached) {
+      return cached.slice();
+    }
     const root = path.substring(0, rootLength);
     const rest = path.substring(rootLength).split(directorySeparator);
     if (rest.length && !lastOrUndefined(rest))
       rest.pop();
-    return [root, ...rest];
+    const res = [root, ...rest];
+    pathComponentsCache.set(path, res);
+    return res;
   }
   function getPathComponents(path, currentDirectory = "") {
     path = combinePaths(currentDirectory, path);
@@ -6972,7 +6981,7 @@ ${lanes.join("\n")}
   function isNodeModulesDirectory(dirPath) {
     return endsWith(dirPath, "/node_modules");
   }
-  var directorySeparator, altDirectorySeparator, urlSchemeSeparator, backslashRegExp, relativePathSegmentRegExp;
+  var directorySeparator, altDirectorySeparator, urlSchemeSeparator, backslashRegExp, pathComponentsCache, relativePathSegmentRegExp;
   var init_path = __esm({
     "src/compiler/path.ts"() {
       "use strict";
@@ -6981,6 +6990,7 @@ ${lanes.join("\n")}
       altDirectorySeparator = "\\";
       urlSchemeSeparator = "://";
       backslashRegExp = /\\/g;
+      pathComponentsCache = /* @__PURE__ */ new Map();
       relativePathSegmentRegExp = /(?:\/\/)|(?:^|\/)\.\.?(?:$|\/)/;
     }
   });
@@ -39112,6 +39122,73 @@ ${lanes.join("\n")}
     }
   });

+  // src/compiler/pnp.ts
+  function getPnpApi(path) {
+    if (typeof process.versions.pnp === "undefined") {
+      return;
+    }
+    const { findPnpApi } =   require("module");
+    if (findPnpApi) {
+      return findPnpApi(`${path}/`);
+    }
+  }
+  function getPnpApiPath(path) {
+    var _a;
+    return (_a = getPnpApi(path)) == null ? void 0 : _a.resolveRequest(
+      "pnpapi",
+      /*issuer*/
+      null
+    );
+  }
+  function getPnpTypeRoots(currentDirectory) {
+    const pnpApi = getPnpApi(currentDirectory);
+    if (!pnpApi) {
+      return [];
+    }
+    currentDirectory = resolvePath(currentDirectory);
+    const currentPackage = pnpApi.findPackageLocator(`${currentDirectory}/`);
+    if (!currentPackage) {
+      return [];
+    }
+    const { packageDependencies } = pnpApi.getPackageInformation(currentPackage);
+    const typeRoots = [];
+    for (const [name, referencish] of Array.from(packageDependencies.entries())) {
+      if (name.startsWith(`@types/`) && referencish !== null) {
+        const dependencyLocator = pnpApi.getLocator(name, referencish);
+        const { packageLocation } = pnpApi.getPackageInformation(dependencyLocator);
+        typeRoots.push(getDirectoryPath(packageLocation));
+      }
+    }
+    return typeRoots;
+  }
+  function isImportablePathPnp(fromPath, toPath3) {
+    const pnpApi = getPnpApi(fromPath);
+    const fromLocator = pnpApi.findPackageLocator(fromPath);
+    const toLocator = pnpApi.findPackageLocator(toPath3);
+    if (toLocator === null) {
+      return false;
+    }
+    const fromInfo = pnpApi.getPackageInformation(fromLocator);
+    const toReference = fromInfo.packageDependencies.get(toLocator.name);
+    if (toReference) {
+      return toReference === toLocator.reference;
+    }
+    for (const reference of fromInfo.packageDependencies.values()) {
+      if (Array.isArray(reference)) {
+        if (reference[0] === toLocator.name && reference[1] === toLocator.reference) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+  var init_pnp = __esm({
+    "src/compiler/pnp.ts"() {
+      "use strict";
+      init_path();
+    }
+  });
+
   // src/compiler/moduleNameResolver.ts
   function trace(host, message, ...args) {
     host.trace(formatMessage(message, ...args));
@@ -39333,7 +39410,7 @@ ${lanes.join("\n")}
       return getDefaultTypeRoots(currentDirectory);
     }
   }
-  function getDefaultTypeRoots(currentDirectory) {
+  function getNodeModulesTypeRoots(currentDirectory) {
     let typeRoots;
     forEachAncestorDirectory(normalizePath(currentDirectory), (directory) => {
       const atTypes = combinePaths(directory, nodeModulesAtTypes);
@@ -39345,6 +39422,15 @@ ${lanes.join("\n")}
     const useCaseSensitiveFileNames2 = typeof host.useCaseSensitiveFileNames === "function" ? host.useCaseSensitiveFileNames() : host.useCaseSensitiveFileNames;
     return comparePaths(path1, path2, !useCaseSensitiveFileNames2) === 0 /* EqualTo */;
   }
+  function getDefaultTypeRoots(currentDirectory) {
+    const nmTypes = getNodeModulesTypeRoots(currentDirectory);
+    const pnpTypes = getPnpTypeRoots(currentDirectory);
+    if (nmTypes == null ? void 0 : nmTypes.length) {
+      return [...nmTypes, ...pnpTypes];
+    } else if (pnpTypes.length) {
+      return pnpTypes;
+    }
+  }
   function getOriginalAndResolvedFileName(fileName, host, traceEnabled) {
     const resolvedFileName = realPath(fileName, host, traceEnabled);
     const pathsAreEqual = arePathsEqual(fileName, resolvedFileName, host);
@@ -39584,6 +39670,21 @@ ${lanes.join("\n")}
   }
   function resolvePackageNameToPackageJson(packageName, containingDirectory, options, host, cache) {
     const moduleResolutionState = getTemporaryModuleResolutionState(cache == null ? void 0 : cache.getPackageJsonInfoCache(), host, options);
+    const pnpapi = getPnpApi(containingDirectory);
+    if (pnpapi) {
+      try {
+        const resolution = pnpapi.resolveToUnqualified(packageName, `${containingDirectory}/`, { considerBuiltins: false });
+        const candidate = normalizeSlashes(resolution).replace(/\/$/, "");
+        return getPackageJsonInfo(
+          candidate,
+          /*onlyRecordFailures*/
+          false,
+          moduleResolutionState
+        );
+      } catch {
+        return;
+      }
+    }
     return forEachAncestorDirectory(containingDirectory, (ancestorDirectory) => {
       if (getBaseFileName(ancestorDirectory) !== "node_modules") {
         const nodeModulesFolder = combinePaths(ancestorDirectory, "node_modules");
@@ -41350,7 +41451,15 @@ ${lanes.join("\n")}
       return lookup(secondaryExtensions);
     }
     function lookup(extensions2) {
-      return forEachAncestorDirectory(normalizeSlashes(directory), (ancestorDirectory) => {
+      const issuer = normalizeSlashes(directory);
+      if (getPnpApi(issuer)) {
+        const resolutionFromCache = tryFindNonRelativeModuleNameInCache(cache, moduleName, mode, issuer, redirectedReference, state);
+        if (resolutionFromCache) {
+          return resolutionFromCache;
+        }
+        return toSearchResult(loadModuleFromImmediateNodeModulesDirectoryPnP(extensions2, moduleName, issuer, state, typesScopeOnly, cache, redirectedReference));
+      }
+      return forEachAncestorDirectory(issuer, (ancestorDirectory) => {
         if (getBaseFileName(ancestorDirectory) !== "node_modules") {
           const resolutionFromCache = tryFindNonRelativeModuleNameInCache(cache, moduleName, mode, ancestorDirectory, redirectedReference, state);
           if (resolutionFromCache) {
@@ -41385,11 +41494,40 @@ ${lanes.join("\n")}
       return loadModuleFromSpecificNodeModulesDirectory(4 /* Declaration */, mangleScopedPackageNameWithTrace(moduleName, state), nodeModulesAtTypes2, nodeModulesAtTypesExists, state, cache, redirectedReference);
     }
   }
+  function loadModuleFromImmediateNodeModulesDirectoryPnP(extensions, moduleName, directory, state, typesScopeOnly, cache, redirectedReference) {
+    const issuer = normalizeSlashes(directory);
+    if (!typesScopeOnly) {
+      const packageResult = tryLoadModuleUsingPnpResolution(extensions, moduleName, issuer, state, cache, redirectedReference);
+      if (packageResult) {
+        return packageResult;
+      }
+    }
+    if (extensions & 4 /* Declaration */) {
+      return tryLoadModuleUsingPnpResolution(4 /* Declaration */, `@types/${mangleScopedPackageNameWithTrace(moduleName, state)}`, issuer, state, cache, redirectedReference);
+    }
+  }
   function loadModuleFromSpecificNodeModulesDirectory(extensions, moduleName, nodeModulesDirectory, nodeModulesDirectoryExists, state, cache, redirectedReference) {
-    var _a, _b;
     const candidate = normalizePath(combinePaths(nodeModulesDirectory, moduleName));
     const { packageName, rest } = parsePackageName(moduleName);
     const packageDirectory = combinePaths(nodeModulesDirectory, packageName);
+    return loadModuleFromSpecificNodeModulesDirectoryImpl(extensions, nodeModulesDirectoryExists, state, cache, redirectedReference, candidate, rest, packageDirectory);
+  }
+  function loadModuleFromPnpResolution(extensions, packageDirectory, rest, state, cache, redirectedReference) {
+    const candidate = normalizePath(combinePaths(packageDirectory, rest));
+    return loadModuleFromSpecificNodeModulesDirectoryImpl(
+      extensions,
+      /*nodeModulesDirectoryExists*/
+      true,
+      state,
+      cache,
+      redirectedReference,
+      candidate,
+      rest,
+      packageDirectory
+    );
+  }
+  function loadModuleFromSpecificNodeModulesDirectoryImpl(extensions, nodeModulesDirectoryExists, state, cache, redirectedReference, candidate, rest, packageDirectory) {
+    var _a, _b;
     let rootPackageInfo;
     let packageInfo = getPackageJsonInfo(candidate, !nodeModulesDirectoryExists, state);
     if (rest !== "" && packageInfo && (!(state.features & 8 /* Exports */) || !hasProperty(((_a = rootPackageInfo = getPackageJsonInfo(packageDirectory, !nodeModulesDirectoryExists, state)) == null ? void 0 : _a.contents.packageJsonContent) ?? emptyArray, "exports"))) {
@@ -41685,11 +41823,30 @@ ${lanes.join("\n")}
   function useCaseSensitiveFileNames(state) {
     return !state.host.useCaseSensitiveFileNames ? true : typeof state.host.useCaseSensitiveFileNames === "boolean" ? state.host.useCaseSensitiveFileNames : state.host.useCaseSensitiveFileNames();
   }
-  var typeScriptVersion, nodeModulesAtTypes, NodeResolutionFeatures, nodeModulesPathPart, mangledScopedPackageSeparator;
+  function loadPnpPackageResolution(packageName, containingDirectory) {
+    if (failedLookups.has(packageName)) {
+      return;
+    }
+    try {
+      const resolution = getPnpApi(containingDirectory).resolveToUnqualified(packageName, `${containingDirectory}/`, { considerBuiltins: false });
+      return normalizeSlashes(resolution).replace(/\/$/, "");
+    } catch {
+      if (isBuiltin(packageName) || packageName.startsWith("@types/") && isBuiltin(packageName.slice(7)) || (packageName.startsWith("@typescript/") || packageName.startsWith("@types/typescript__"))) {
+        failedLookups.add(packageName);
+      }
+    }
+  }
+  function tryLoadModuleUsingPnpResolution(extensions, moduleName, containingDirectory, state, cache, redirectedReference) {
+    const { packageName, rest } = parsePackageName(moduleName);
+    const packageResolution = loadPnpPackageResolution(packageName, containingDirectory);
+    return packageResolution ? loadModuleFromPnpResolution(extensions, packageResolution, rest, state, cache, redirectedReference) : void 0;
+  }
+  var typeScriptVersion, nodeModulesAtTypes, NodeResolutionFeatures, nodeModulesPathPart, mangledScopedPackageSeparator, isBuiltin, failedLookups;
   var init_moduleNameResolver = __esm({
     "src/compiler/moduleNameResolver.ts"() {
       "use strict";
       init_ts2();
+      init_pnp();
       nodeModulesAtTypes = combinePaths("node_modules", "@types");
       NodeResolutionFeatures = /* @__PURE__ */ ((NodeResolutionFeatures2) => {
         NodeResolutionFeatures2[NodeResolutionFeatures2["None"] = 0] = "None";
@@ -41706,6 +41863,8 @@ ${lanes.join("\n")}
       })(NodeResolutionFeatures || {});
       nodeModulesPathPart = "/node_modules/";
       mangledScopedPackageSeparator = "__";
+      ({ isBuiltin } =   require("module"));
+      failedLookups = /* @__PURE__ */ new Set();
     }
   });

@@ -45064,7 +45223,15 @@ ${lanes.join("\n")}
       /*preferSymlinks*/
       true,
       (path, isRedirect) => {
-        const isInNodeModules = pathContainsNodeModules(path);
+        let isInNodeModules = pathContainsNodeModules(path);
+        const pnpapi = getPnpApi(path);
+        if (!isInNodeModules && pnpapi) {
+          const fromLocator = pnpapi.findPackageLocator(importingFileName);
+          const toLocator = pnpapi.findPackageLocator(path);
+          if (fromLocator && toLocator && fromLocator !== toLocator) {
+            isInNodeModules = true;
+          }
+        }
         allFileNames.set(path, { path: getCanonicalFileName(path), isRedirect, isInNodeModules });
         importedFileFromNodeModules = importedFileFromNodeModules || isInNodeModules;
       }
@@ -45255,7 +45422,41 @@ ${lanes.join("\n")}
     if (!host.fileExists || !host.readFile) {
       return void 0;
     }
-    const parts = getNodeModulePathParts(path);
+    let parts = getNodeModulePathParts(path);
+    let pnpPackageName;
+    const pnpApi = getPnpApi(path);
+    if (pnpApi) {
+      const fromLocator = pnpApi.findPackageLocator(importingSourceFile.fileName);
+      const toLocator = pnpApi.findPackageLocator(path);
+      if (fromLocator === toLocator) {
+        return void 0;
+      }
+      if (fromLocator && toLocator) {
+        const fromInfo = pnpApi.getPackageInformation(fromLocator);
+        if (toLocator.reference === fromInfo.packageDependencies.get(toLocator.name)) {
+          pnpPackageName = toLocator.name;
+        } else {
+          for (const [name, reference] of fromInfo.packageDependencies) {
+            if (Array.isArray(reference)) {
+              if (reference[0] === toLocator.name && reference[1] === toLocator.reference) {
+                pnpPackageName = name;
+                break;
+              }
+            }
+          }
+        }
+        if (!parts) {
+          const toInfo = pnpApi.getPackageInformation(toLocator);
+          parts = {
+            topLevelNodeModulesIndex: void 0,
+            topLevelPackageNameIndex: void 0,
+            // The last character from packageLocation is the trailing "/", we want to point to it
+            packageRootIndex: toInfo.packageLocation.length - 1,
+            fileNameIndex: path.lastIndexOf(`/`)
+          };
+        }
+      }
+    }
     if (!parts) {
       return void 0;
     }
@@ -45293,14 +45494,16 @@ ${lanes.join("\n")}
     if (isRedirect && !isPackageRootPath) {
       return void 0;
     }
-    const globalTypingsCacheLocation = host.getGlobalTypingsCacheLocation && host.getGlobalTypingsCacheLocation();
-    const pathToTopLevelNodeModules = getCanonicalFileName(moduleSpecifier.substring(0, parts.topLevelNodeModulesIndex));
-    if (!(startsWith(sourceDirectory, pathToTopLevelNodeModules) || globalTypingsCacheLocation && startsWith(getCanonicalFileName(globalTypingsCacheLocation), pathToTopLevelNodeModules))) {
-      return void 0;
+    if (typeof process.versions.pnp === "undefined") {
+      const globalTypingsCacheLocation = host.getGlobalTypingsCacheLocation && host.getGlobalTypingsCacheLocation();
+      const pathToTopLevelNodeModules = getCanonicalFileName(moduleSpecifier.substring(0, parts.topLevelNodeModulesIndex));
+      if (!(startsWith(sourceDirectory, pathToTopLevelNodeModules) || globalTypingsCacheLocation && startsWith(getCanonicalFileName(globalTypingsCacheLocation), pathToTopLevelNodeModules))) {
+        return void 0;
+      }
     }
-    const nodeModulesDirectoryName = moduleSpecifier.substring(parts.topLevelPackageNameIndex + 1);
-    const packageName = getPackageNameFromTypesPackageName(nodeModulesDirectoryName);
-    return getEmitModuleResolutionKind(options) === 1 /* Classic */ && packageName === nodeModulesDirectoryName ? void 0 : packageName;
+    const nodeModulesDirectoryName = typeof pnpPackageName !== "undefined" ? pnpPackageName + moduleSpecifier.substring(parts.packageRootIndex) : moduleSpecifier.substring(parts.topLevelPackageNameIndex + 1);
+    const packageNameFromPath = getPackageNameFromTypesPackageName(nodeModulesDirectoryName);
+    return getEmitModuleResolutionKind(options) === 1 /* Classic */ && packageNameFromPath === nodeModulesDirectoryName ? void 0 : packageNameFromPath;
     function tryDirectoryWithPackageJson(packageRootIndex) {
       var _a, _b;
       const packageRootPath = path.substring(0, packageRootIndex);
@@ -45313,9 +45516,9 @@ ${lanes.join("\n")}
         const importMode = overrideMode || importingSourceFile.impliedNodeFormat;
         if (getResolvePackageJsonExports(options)) {
           const nodeModulesDirectoryName2 = packageRootPath.substring(parts.topLevelPackageNameIndex + 1);
-          const packageName2 = getPackageNameFromTypesPackageName(nodeModulesDirectoryName2);
+          const packageName = getPackageNameFromTypesPackageName(pnpPackageName ? pnpPackageName : nodeModulesDirectoryName2);
           const conditions = getConditions(options, importMode === 99 /* ESNext */);
-          const fromExports = packageJsonContent.exports ? tryGetModuleNameFromExports(options, path, packageRootPath, packageName2, packageJsonContent.exports, conditions) : void 0;
+          const fromExports = packageJsonContent.exports ? tryGetModuleNameFromExports(options, path, packageRootPath, packageName, packageJsonContent.exports, conditions) : void 0;
           if (fromExports) {
             const withJsExtension = !hasTSFileExtension(fromExports.moduleFileToTry) ? fromExports : { moduleFileToTry: removeFileExtension(fromExports.moduleFileToTry) + tryGetJSExtensionForFile(fromExports.moduleFileToTry, options) };
             return { ...withJsExtension, verbatimFromExports: true };
@@ -45467,6 +45670,7 @@ ${lanes.join("\n")}
     "src/compiler/moduleSpecifiers.ts"() {
       "use strict";
       init_ts2();
+      init_pnp();
     }
   });

@@ -47944,7 +48148,7 @@ ${lanes.join("\n")}
       }
     }
     function getTargetofModuleDefault(moduleSymbol, node, dontResolveAlias) {
-      var _a;
+      var _a, _b;
       let exportDefaultSymbol;
       if (isShorthandAmbientModuleSymbol(moduleSymbol)) {
         exportDefaultSymbol = moduleSymbol;
@@ -47953,7 +48157,7 @@ ${lanes.join("\n")}
       }
       const file = (_a = moduleSymbol.declarations) == null ? void 0 : _a.find(isSourceFile);
       const specifier = getModuleSpecifierForImportOrExport(node);
-      if (!specifier) {
+      if (!specifier || moduleSymbol.flags === 512 && ((_b = moduleSymbol.exports) == null ? void 0 : _b.has("default"))) {
         return exportDefaultSymbol;
       }
       const hasDefaultOnly = isOnlyImportedAsDefault(specifier);
@@ -111946,10 +112150,11 @@ ${lanes.join("\n")}
     return fileExtensionIs(fileName, ".json" /* Json */) ? ".json" /* Json */ : options.jsx === 1 /* Preserve */ && fileExtensionIsOneOf(fileName, [".jsx" /* Jsx */, ".tsx" /* Tsx */]) ? ".jsx" /* Jsx */ : fileExtensionIsOneOf(fileName, [".mts" /* Mts */, ".mjs" /* Mjs */]) ? ".mjs" /* Mjs */ : fileExtensionIsOneOf(fileName, [".cts" /* Cts */, ".cjs" /* Cjs */]) ? ".cjs" /* Cjs */ : ".js" /* Js */;
   }
   function getOutputPathWithoutChangingExt(inputFileName, configFile, ignoreCase, outputDir, getCommonSourceDirectory2) {
-    return outputDir ? resolvePath(
-      outputDir,
-      getRelativePathFromDirectory(getCommonSourceDirectory2 ? getCommonSourceDirectory2() : getCommonSourceDirectoryOfConfig(configFile, ignoreCase), inputFileName, ignoreCase)
-    ) : inputFileName;
+    if (outputDir) {
+      const commonDir = getCommonSourceDirectory2 ? getCommonSourceDirectory2() : getCommonSourceDirectoryOfConfig(configFile, ignoreCase);
+      return `${outputDir}${directorySeparator}${inputFileName.slice(commonDir.length)}`;
+    }
+    return inputFileName;
   }
   function getOutputDeclarationFileName(inputFileName, configFile, ignoreCase, getCommonSourceDirectory2) {
     return changeExtension(
@@ -117865,6 +118070,7 @@ ${lanes.join("\n")}
         ConfigFileProgramReloadLevel2[ConfigFileProgramReloadLevel2["None"] = 0] = "None";
         ConfigFileProgramReloadLevel2[ConfigFileProgramReloadLevel2["Partial"] = 1] = "Partial";
         ConfigFileProgramReloadLevel2[ConfigFileProgramReloadLevel2["Full"] = 2] = "Full";
+        ConfigFileProgramReloadLevel2[ConfigFileProgramReloadLevel2["Resolutions"] = 3] = "Resolutions";
         return ConfigFileProgramReloadLevel2;
       })(ConfigFileProgramReloadLevel || {});
       WatchLogLevel = /* @__PURE__ */ ((WatchLogLevel2) => {
@@ -125506,6 +125712,11 @@ ${lanes.join("\n")}
     if (configFileName) {
       configFileWatcher = watchFile2(configFileName, scheduleProgramReload, 2e3 /* High */, watchOptions, WatchType.ConfigFile);
     }
+    let pnpFileWatcher;
+    const pnpApiPath = getPnpApiPath(__filename);
+    if (pnpApiPath) {
+      pnpFileWatcher = watchFile2(pnpApiPath, scheduleResolutionReload, 2e3 /* High */, watchOptions, WatchType.ConfigFile);
+    }
     const compilerHost = createCompilerHostFromProgramHost(host, () => compilerOptions, directoryStructureHost);
     setGetSourceFileAsHashVersioned(compilerHost);
     const getNewSourceFile = compilerHost.getSourceFile;
@@ -125569,6 +125780,10 @@ ${lanes.join("\n")}
         configFileWatcher.close();
         configFileWatcher = void 0;
       }
+      if (pnpFileWatcher) {
+        pnpFileWatcher.close();
+        pnpFileWatcher = void 0;
+      }
       extendedConfigCache == null ? void 0 : extendedConfigCache.clear();
       extendedConfigCache = void 0;
       if (sharedExtendedConfigFileWatchers) {
@@ -125604,7 +125819,7 @@ ${lanes.join("\n")}
     function getCurrentProgram() {
       return builderProgram && builderProgram.getProgramOrUndefined();
     }
-    function synchronizeProgram() {
+    function synchronizeProgram(forceAllFilesAsInvalidated = false) {
       writeLog(`Synchronizing program`);
       Debug.assert(compilerOptions);
       Debug.assert(rootFileNames);
@@ -125616,7 +125831,7 @@ ${lanes.join("\n")}
           resolutionCache.onChangesAffectModuleResolution();
         }
       }
-      const { hasInvalidatedResolutions, hasInvalidatedLibResolutions } = resolutionCache.createHasInvalidatedResolutions(customHasInvalidatedResolutions, customHasInvalidLibResolutions);
+      const { hasInvalidatedResolutions, hasInvalidatedLibResolutions } = resolutionCache.createHasInvalidatedResolutions(forceAllFilesAsInvalidated ? returnTrue : customHasInvalidatedResolutions, customHasInvalidLibResolutions);
       const {
         originalReadFile,
         originalFileExists,
@@ -125824,13 +126039,19 @@ ${lanes.join("\n")}
       reloadLevel = 2 /* Full */;
       scheduleProgramUpdate();
     }
+    function scheduleResolutionReload() {
+      writeLog("Clearing resolutions");
+      resolutionCache.clear();
+      reloadLevel = 3 /* Resolutions */;
+      scheduleProgramUpdate();
+    }
     function updateProgramWithWatchStatus() {
       timerToUpdateProgram = void 0;
       reportFileChangeDetectedOnCreateProgram = true;
       updateProgram();
     }
     function updateProgram() {
-      var _a, _b, _c, _d;
+      var _a, _b, _c, _d, _e;
       switch (reloadLevel) {
         case 1 /* Partial */:
           (_a = perfLogger) == null ? void 0 : _a.logStartUpdateProgram("PartialConfigReload");
@@ -125840,12 +126061,19 @@ ${lanes.join("\n")}
           (_b = perfLogger) == null ? void 0 : _b.logStartUpdateProgram("FullConfigReload");
           reloadConfigFile();
           break;
+        case 3 /* Resolutions */:
+          (_c = perfLogger) == null ? void 0 : _c.logStartUpdateProgram("SynchronizeProgramWithResolutions");
+          synchronizeProgram(
+            /*forceAllFilesAsInvalidated*/
+            true
+          );
+          break;
         default:
-          (_c = perfLogger) == null ? void 0 : _c.logStartUpdateProgram("SynchronizeProgram");
+          (_d = perfLogger) == null ? void 0 : _d.logStartUpdateProgram("SynchronizeProgram");
           synchronizeProgram();
           break;
       }
-      (_d = perfLogger) == null ? void 0 : _d.logStopUpdateProgram("Done");
+      (_e = perfLogger) == null ? void 0 : _e.logStopUpdateProgram("Done");
       return getCurrentBuilderProgram();
     }
     function reloadFileNamesFromConfigFile() {
@@ -126133,6 +126361,7 @@ ${lanes.join("\n")}
     "src/compiler/watchPublic.ts"() {
       "use strict";
       init_ts2();
+      init_pnp();
     }
   });

@@ -131435,7 +131664,14 @@ ${lanes.join("\n")}
       return false;
     }
     function getNodeModulesPackageNameFromFileName(importedFileName, moduleSpecifierResolutionHost) {
-      if (!importedFileName.includes("node_modules")) {
+      const pnpapi = getPnpApi(importedFileName);
+      if (pnpapi) {
+        const fromLocator = pnpapi.findPackageLocator(fromFile.fileName);
+        const toLocator = pnpapi.findPackageLocator(importedFileName);
+        if (!(fromLocator && toLocator)) {
+          return void 0;
+        }
+      } else if (!importedFileName.includes("node_modules")) {
         return void 0;
       }
       const specifier = ts_moduleSpecifiers_exports.getNodeModulesPackageName(
@@ -131731,6 +131967,7 @@ ${lanes.join("\n")}
   var init_utilities4 = __esm({
     "src/services/utilities.ts"() {
       "use strict";
+      init_pnp();
       init_ts4();
       scanner = createScanner(
         99 /* Latest */,
@@ -132009,6 +132246,9 @@ ${lanes.join("\n")}
     return hasImportablePath;
   }
   function isImportablePath(fromPath, toPath3, getCanonicalFileName, globalCachePath) {
+    if (getPnpApi(fromPath)) {
+      return isImportablePathPnp(fromPath, toPath3);
+    }
     const toNodeModules = forEachAncestorDirectory(toPath3, (ancestor) => getBaseFileName(ancestor) === "node_modules" ? ancestor : void 0);
     const toNodeModulesParent = toNodeModules && getDirectoryPath(getCanonicalFileName(toNodeModules));
     return toNodeModulesParent === void 0 || startsWith(getCanonicalFileName(fromPath), toNodeModulesParent) || !!globalCachePath && startsWith(getCanonicalFileName(globalCachePath), toNodeModulesParent);
@@ -132192,6 +132432,7 @@ ${lanes.join("\n")}
   var init_exportInfoMap = __esm({
     "src/services/exportInfoMap.ts"() {
       "use strict";
+      init_pnp();
       init_ts4();
       ImportKind = /* @__PURE__ */ ((ImportKind2) => {
         ImportKind2[ImportKind2["Named"] = 0] = "Named";
@@ -161005,7 +161246,31 @@ ${newComment.split("\n").map((c) => ` * ${c}`).join("\n")}
             );
           }
         };
-        if (fragmentDirectory && getResolvePackageJsonExports(compilerOptions)) {
+        const checkExports = (packageFile, packageDirectory, fragmentSubpath) => {
+          const packageJson = readJson(packageFile, host);
+          const exports = packageJson.exports;
+          if (exports) {
+            if (typeof exports !== "object" || exports === null) {
+              return true;
+            }
+            const keys = getOwnKeys(exports);
+            const conditions = getConditions(compilerOptions, mode === 99 /* ESNext */);
+            addCompletionEntriesFromPathsOrExports(
+              result,
+              fragmentSubpath,
+              packageDirectory,
+              extensionOptions,
+              host,
+              keys,
+              (key) => singleElementArray(getPatternFromFirstMatchingCondition(exports[key], conditions)),
+              comparePatternKeys
+            );
+            return true;
+          }
+          return false;
+        };
+        const shouldCheckExports = fragmentDirectory && getResolvePackageJsonExports(compilerOptions);
+        if (shouldCheckExports) {
           const nodeModulesDirectoryLookup = ancestorLookup;
           ancestorLookup = (ancestor) => {
             const components = getPathComponents(fragment);
@@ -161024,32 +161289,52 @@ ${newComment.split("\n").map((c) => ` * ${c}`).join("\n")}
             const packageDirectory = combinePaths(ancestor, "node_modules", packagePath);
             const packageFile = combinePaths(packageDirectory, "package.json");
             if (tryFileExists(host, packageFile)) {
-              const packageJson = readJson(packageFile, host);
-              const exports = packageJson.exports;
-              if (exports) {
-                if (typeof exports !== "object" || exports === null) {
-                  return;
+              const fragmentSubpath = components.join("/") + (components.length && hasTrailingDirectorySeparator(fragment) ? "/" : "");
+              if (checkExports(packageFile, packageDirectory, fragmentSubpath)) {
+                return;
+              }
+            }
+            return nodeModulesDirectoryLookup(ancestor);
+          };
+        }
+        const pnpApi = getPnpApi(scriptPath);
+        if (pnpApi) {
+          const pathRegExp = /^(?![a-zA-Z]:[\\/]|\\\\|\.{0,2}(?:\/|$))((?:@[^/]+\/)?[^/]+)\/*(.*|)$/;
+          const dependencyNameMatch = fragment.match(pathRegExp);
+          if (dependencyNameMatch) {
+            const [, dependencyName, subPath] = dependencyNameMatch;
+            let unqualified;
+            try {
+              unqualified = pnpApi.resolveToUnqualified(dependencyName, scriptPath, { considerBuiltins: false });
+            } catch {
+            }
+            if (unqualified) {
+              const packageDirectory = normalizePath(unqualified);
+              let shouldGetCompletions = true;
+              if (shouldCheckExports) {
+                const packageFile = combinePaths(packageDirectory, "package.json");
+                if (tryFileExists(host, packageFile) && checkExports(packageFile, packageDirectory, subPath)) {
+                  shouldGetCompletions = false;
                 }
-                const keys = getOwnKeys(exports);
-                const fragmentSubpath = components.join("/") + (components.length && hasTrailingDirectorySeparator(fragment) ? "/" : "");
-                const conditions = getConditions(compilerOptions, mode === 99 /* ESNext */);
-                addCompletionEntriesFromPathsOrExports(
-                  result,
-                  fragmentSubpath,
+              }
+              if (shouldGetCompletions) {
+                getCompletionEntriesForDirectoryFragment(
+                  subPath,
                   packageDirectory,
                   extensionOptions,
                   host,
-                  keys,
-                  (key) => singleElementArray(getPatternFromFirstMatchingCondition(exports[key], conditions)),
-                  comparePatternKeys
+                  /*moduleSpecifierIsRelative*/
+                  false,
+                  /*exclude*/
+                  void 0,
+                  result
                 );
-                return;
               }
             }
-            return nodeModulesDirectoryLookup(ancestor);
-          };
+          }
+        } else {
+          forEachAncestorDirectory(scriptPath, ancestorLookup);
         }
-        forEachAncestorDirectory(scriptPath, ancestorLookup);
       }
     }
     return arrayFrom(result.values());
@@ -161180,9 +161465,15 @@ ${newComment.split("\n").map((c) => ` * ${c}`).join("\n")}
     for (const root of typeRoots) {
       getCompletionEntriesFromDirectories(root);
     }
-    for (const packageJson of findPackageJsons(scriptPath, host)) {
-      const typesDir = combinePaths(getDirectoryPath(packageJson), "node_modules/@types");
-      getCompletionEntriesFromDirectories(typesDir);
+    if (getPnpApi(scriptPath)) {
+      for (const root of getPnpTypeRoots(scriptPath)) {
+        getCompletionEntriesFromDirectories(root);
+      }
+    } else {
+      for (const packageJson of findPackageJsons(scriptPath, host)) {
+        const typesDir = combinePaths(getDirectoryPath(packageJson), "node_modules/@types");
+        getCompletionEntriesFromDirectories(typesDir);
+      }
     }
     return result;
     function getCompletionEntriesFromDirectories(directory) {
@@ -161265,6 +161556,7 @@ ${newComment.split("\n").map((c) => ` * ${c}`).join("\n")}
   var init_stringCompletions = __esm({
     "src/services/stringCompletions.ts"() {
       "use strict";
+      init_pnp();
       init_ts4();
       init_ts_Completions();
       kindPrecedence = {
@@ -173951,6 +174243,7 @@ ${options.prefix}` : "\n" : options.prefix
   var init_project = __esm({
     "src/server/project.ts"() {
       "use strict";
+      init_pnp();
       init_ts7();
       init_ts7();
       init_ts_server3();
@@ -176029,6 +176322,29 @@ ${options.prefix}` : "\n" : options.prefix
           return this.projectReferences;
         }
         updateReferences(refs) {
+          if (typeof process.versions.pnp !== `undefined`) {
+            const basePath = this.getCurrentDirectory();
+            const getPnpPath = (path) => {
+              try {
+                const pnpApi = getPnpApi(`${path}/`);
+                if (!pnpApi) {
+                  return path;
+                }
+                const targetLocator = pnpApi.findPackageLocator(`${path}/`);
+                const { packageLocation } = pnpApi.getPackageInformation(targetLocator);
+                const request = combinePaths(targetLocator.name, getRelativePathFromDirectory(
+                  packageLocation,
+                  path,
+                  /*ignoreCase*/
+                  false
+                ));
+                return pnpApi.resolveToUnqualified(request, `${basePath}/`);
+              } catch {
+                return path;
+              }
+            };
+            refs = refs == null ? void 0 : refs.map((r) => ({ ...r, path: getPnpPath(r.path) }));
+          }
           this.projectReferences = refs;
           this.potentialProjectReferences = void 0;
         }
@@ -176494,6 +176810,7 @@ ${options.prefix}` : "\n" : options.prefix
   var init_editorServices = __esm({
     "src/server/editorServices.ts"() {
       "use strict";
+      init_pnp();
       init_ts7();
       init_ts_server3();
       init_protocol();
@@ -176721,6 +177038,7 @@ ${options.prefix}` : "\n" : options.prefix
             log,
             getDetailWatchInfo
           );
+          this.pnpWatcher = this.watchPnpFile();
           (_a = opts.incrementalVerifier) == null ? void 0 : _a.call(opts, this);
         }
         toPath(fileName) {
@@ -178532,7 +178850,7 @@ Dynamic files must always be opened with service's current directory or service
           this.performanceEventHandler = performanceEventHandler;
         }
         setHostConfiguration(args) {
-          var _a;
+          var _a, _b;
           if (args.file) {
             const info = this.getScriptInfoForNormalizedPath(toNormalizedPath(args.file));
             if (info) {
@@ -178575,7 +178893,8 @@ Dynamic files must always be opened with service's current directory or service
             }
             if (args.watchOptions) {
               this.hostConfiguration.watchOptions = (_a = convertWatchOptions(args.watchOptions)) == null ? void 0 : _a.watchOptions;
-              this.logger.info(`Host watch options changed to ${JSON.stringify(this.hostConfiguration.watchOptions)}, it will be take effect for next watches.`);
+              (_b = this.pnpWatcher) == null ? void 0 : _b.close();
+              this.watchPnpFile();
             }
           }
         }
@@ -179522,6 +179841,29 @@ Dynamic files must always be opened with service's current directory or service
           });
         }
         /** @internal */
+        watchPnpFile() {
+          const pnpApiPath = getPnpApiPath(__filename);
+          if (!pnpApiPath) {
+            return;
+          }
+          return this.watchFactory.watchFile(
+            pnpApiPath,
+            () => {
+              this.forEachProject((project) => {
+                for (const info of project.getScriptInfos()) {
+                  project.resolutionCache.invalidateResolutionOfFile(info.path);
+                }
+                project.markAsDirty();
+                updateProjectIfDirty(project);
+              });
+              this.delayEnsureProjectForOpenFiles();
+            },
+            250 /* Low */,
+            this.hostConfiguration.watchOptions,
+            WatchType.ConfigFile
+          );
+        }
+        /** @internal */
         watchPackageJsonFile(path) {
           const watchers = this.packageJsonFilesMap || (this.packageJsonFilesMap = /* @__PURE__ */ new Map());
           if (!watchers.has(path)) {
diff --git a/lib/typingsInstaller.js b/lib/typingsInstaller.js
index 0692c1e7b..97d2d1b08 100644
semver exclusivity >=5.3.0
--- a/lib/typingsInstaller.js
+++ b/lib/typingsInstaller.js
@@ -4459,6 +4459,9 @@ var sys = (() => {
       return Buffer2.from && Buffer2.from !== Int8Array.from ? Buffer2.from(input, encoding) : new Buffer2(input, encoding);
     }
     function isFileSystemCaseSensitive() {
+      if (typeof process.versions.pnp !== `undefined`) {
+        return true;
+      }
       if (platform === "win32" || platform === "win64") {
         return false;
       }
@@ -4833,12 +4836,19 @@ function getAnyExtensionFromPath(path2, extensions, ignoreCase) {
   }
   return "";
 }
+var pathComponentsCache = /* @__PURE__ */ new Map();
 function pathComponents(path2, rootLength) {
+  const cached = pathComponentsCache.get(path2);
+  if (cached) {
+    return cached.slice();
+  }
   const root = path2.substring(0, rootLength);
   const rest = path2.substring(rootLength).split(directorySeparator);
   if (rest.length && !lastOrUndefined(rest))
     rest.pop();
-  return [root, ...rest];
+  const res = [root, ...rest];
+  pathComponentsCache.set(path2, res);
+  return res;
 }
 function getPathComponents(path2, currentDirectory = "") {
   path2 = combinePaths(currentDirectory, path2);
@@ -27280,6 +27290,17 @@ function specToDiagnostic(spec, disallowTrailingRecursion) {
   }
 }

+// src/compiler/pnp.ts
+function getPnpApi(path2) {
+  if (typeof process.versions.pnp === "undefined") {
+    return;
+  }
+  const { findPnpApi } = require("module");
+  if (findPnpApi) {
+    return findPnpApi(`${path2}/`);
+  }
+}
+
 // src/compiler/moduleNameResolver.ts
 function trace(host, message, ...args) {
   host.trace(formatMessage(message, ...args));
@@ -28751,7 +28772,15 @@ function loadModuleFromNearestNodeModulesDirectoryWorker(extensions, moduleName,
     return lookup(secondaryExtensions);
   }
   function lookup(extensions2) {
-    return forEachAncestorDirectory(normalizeSlashes(directory), (ancestorDirectory) => {
+    const issuer = normalizeSlashes(directory);
+    if (getPnpApi(issuer)) {
+      const resolutionFromCache = tryFindNonRelativeModuleNameInCache(cache, moduleName, mode, issuer, redirectedReference, state);
+      if (resolutionFromCache) {
+        return resolutionFromCache;
+      }
+      return toSearchResult(loadModuleFromImmediateNodeModulesDirectoryPnP(extensions2, moduleName, issuer, state, typesScopeOnly, cache, redirectedReference));
+    }
+    return forEachAncestorDirectory(issuer, (ancestorDirectory) => {
       if (getBaseFileName(ancestorDirectory) !== "node_modules") {
         const resolutionFromCache = tryFindNonRelativeModuleNameInCache(cache, moduleName, mode, ancestorDirectory, redirectedReference, state);
         if (resolutionFromCache) {
@@ -28786,11 +28815,40 @@ function loadModuleFromImmediateNodeModulesDirectory(extensions, moduleName, dir
     return loadModuleFromSpecificNodeModulesDirectory(4 /* Declaration */, mangleScopedPackageNameWithTrace(moduleName, state), nodeModulesAtTypes2, nodeModulesAtTypesExists, state, cache, redirectedReference);
   }
 }
+function loadModuleFromImmediateNodeModulesDirectoryPnP(extensions, moduleName, directory, state, typesScopeOnly, cache, redirectedReference) {
+  const issuer = normalizeSlashes(directory);
+  if (!typesScopeOnly) {
+    const packageResult = tryLoadModuleUsingPnpResolution(extensions, moduleName, issuer, state, cache, redirectedReference);
+    if (packageResult) {
+      return packageResult;
+    }
+  }
+  if (extensions & 4 /* Declaration */) {
+    return tryLoadModuleUsingPnpResolution(4 /* Declaration */, `@types/${mangleScopedPackageNameWithTrace(moduleName, state)}`, issuer, state, cache, redirectedReference);
+  }
+}
 function loadModuleFromSpecificNodeModulesDirectory(extensions, moduleName, nodeModulesDirectory, nodeModulesDirectoryExists, state, cache, redirectedReference) {
-  var _a, _b;
   const candidate = normalizePath(combinePaths(nodeModulesDirectory, moduleName));
   const { packageName, rest } = parsePackageName(moduleName);
   const packageDirectory = combinePaths(nodeModulesDirectory, packageName);
+  return loadModuleFromSpecificNodeModulesDirectoryImpl(extensions, nodeModulesDirectoryExists, state, cache, redirectedReference, candidate, rest, packageDirectory);
+}
+function loadModuleFromPnpResolution(extensions, packageDirectory, rest, state, cache, redirectedReference) {
+  const candidate = normalizePath(combinePaths(packageDirectory, rest));
+  return loadModuleFromSpecificNodeModulesDirectoryImpl(
+    extensions,
+    /*nodeModulesDirectoryExists*/
+    true,
+    state,
+    cache,
+    redirectedReference,
+    candidate,
+    rest,
+    packageDirectory
+  );
+}
+function loadModuleFromSpecificNodeModulesDirectoryImpl(extensions, nodeModulesDirectoryExists, state, cache, redirectedReference, candidate, rest, packageDirectory) {
+  var _a, _b;
   let rootPackageInfo;
   let packageInfo = getPackageJsonInfo(candidate, !nodeModulesDirectoryExists, state);
   if (rest !== "" && packageInfo && (!(state.features & 8 /* Exports */) || !hasProperty(((_a = rootPackageInfo = getPackageJsonInfo(packageDirectory, !nodeModulesDirectoryExists, state)) == null ? void 0 : _a.contents.packageJsonContent) ?? emptyArray, "exports"))) {
@@ -29027,6 +29085,26 @@ function traceIfEnabled(state, diagnostic, ...args) {
 function useCaseSensitiveFileNames(state) {
   return !state.host.useCaseSensitiveFileNames ? true : typeof state.host.useCaseSensitiveFileNames === "boolean" ? state.host.useCaseSensitiveFileNames : state.host.useCaseSensitiveFileNames();
 }
+var { isBuiltin } = require("module");
+var failedLookups = /* @__PURE__ */ new Set();
+function loadPnpPackageResolution(packageName, containingDirectory) {
+  if (failedLookups.has(packageName)) {
+    return;
+  }
+  try {
+    const resolution = getPnpApi(containingDirectory).resolveToUnqualified(packageName, `${containingDirectory}/`, { considerBuiltins: false });
+    return normalizeSlashes(resolution).replace(/\/$/, "");
+  } catch {
+    if (isBuiltin(packageName) || packageName.startsWith("@types/") && isBuiltin(packageName.slice(7)) || (packageName.startsWith("@typescript/") || packageName.startsWith("@types/typescript__"))) {
+      failedLookups.add(packageName);
+    }
+  }
+}
+function tryLoadModuleUsingPnpResolution(extensions, moduleName, containingDirectory, state, cache, redirectedReference) {
+  const { packageName, rest } = parsePackageName(moduleName);
+  const packageResolution = loadPnpPackageResolution(packageName, containingDirectory);
+  return packageResolution ? loadModuleFromPnpResolution(extensions, packageResolution, rest, state, cache, redirectedReference) : void 0;
+}

 // src/compiler/checker.ts
 var nextNodeId = 1;
